<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        <meta name="author" content="shenwei356">
        
        <link rel="shortcut icon" href="../img/favicon.ico">

	<title>Benchmark - fakit - FASTA kit</title>

        <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="../css/font-awesome-4.5.0.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.css">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

            ga('create', 'UA-28948626-5', 'shenwei356.github.io/fakit');
            ga('send', 'pageview');
        </script>
        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="..">fakit - FASTA kit</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="../download/">Download</a>
                    </li>
                
                
                
                    <li >
                        <a href="../usage/">Usage</a>
                    </li>
                
                
                
                    <li class="active">
                        <a href="./">Benchmark</a>
                    </li>
                
                
                
                    <li >
                        <a href="../extension/">Extension</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li >
                        <a rel="next" href="../usage/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../extension/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/shenwei356/fakit">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            
                <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#benchmark">Benchmark</a></li>
        
            <li><a href="#softwares">Softwares</a></li>
        
            <li><a href="#features">Features</a></li>
        
            <li><a href="#datasets">Datasets</a></li>
        
            <li><a href="#platform">Platform</a></li>
        
            <li><a href="#test-1-reverse-complement">Test 1. Reverse Complement</a></li>
        
            <li><a href="#test-2extract-sequencs-by-id-list">Test 2.Extract sequencs by ID list</a></li>
        
            <li><a href="#test-3-deduplication">Test 3. Deduplication</a></li>
        
            <li><a href="#test-4-sampling">Test 4. Sampling</a></li>
        
            <li><a href="#test-5-spliting">Test 5. Spliting</a></li>
        
    
    </ul>
</div></div>
                <div class="col-md-9" role="main">

<h1 id="benchmark">Benchmark</h1>
<h2 id="softwares">Softwares</h2>
<ol>
<li><a href="https://github.com/shenwei356/fakit">fakit</a>. (Go).
   Version <a href="https://github.com/shenwei356/fakit/releases/tag/v0.1.3">v0.1.3</a>.</li>
<li><a href="https://github.com/jimhester/fasta_utilities">fasta_utilities</a>. (Perl).
   Version
   <a href="https://github.com/jimhester/fasta_utilities/commit/329f7ca9266d4a0a96cb5576c464c1bd106865a0">329f7ca</a>.
   <em>Lots of dependencies to install</em>.</li>
<li><a href="http://hannonlab.cshl.edu/fastx_toolkit/">fastx_toolkit</a>. (Perl).
   Version <a href="http://hannonlab.cshl.edu/fastx_toolkit/fastx_toolkit_0.0.13_binaries_Linux_2.6_amd64.tar.bz2">0.0.13</a>.
   <em>Can't handle multi-line FASTA files</em>.</li>
<li><a href="https://github.com/mdshw5/pyfaidx">pyfaidx</a>. (Python).
   Version <a href="https://pypi.python.org/packages/source/p/pyfaidx/pyfaidx-0.4.7.1.tar.gz#md5=f33604a3550c2fa115ac7d33b952127d">0.4.7.1</a>.</li>
<li><a href="http://seqmagick.readthedocs.org/en/latest/index.html">seqmagick</a>. (Python).
   Version 0.6.1</li>
<li><a href="https://github.com/lh3/seqtk">seqtk</a>. (C).
   Version <a href="https://github.com/lh3/seqtk/commit/4feb6e81444ab6bc44139dd3a125068f81ae4ad8">1.0-r82-dirty</a>.</li>
</ol>
<h2 id="features">Features</h2>
<table>
<thead>
<tr>
<th align="left">Features</th>
<th align="center">fakit</th>
<th align="center">fasta_utilities</th>
<th align="center">fastx_toolkit</th>
<th align="center">pyfaidx</th>
<th align="center">seqmagick</th>
<th align="left">seqtk</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Cross platform</td>
<td align="center">Yes</td>
<td align="center">Partly</td>
<td align="center">Partly</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="left">Yes</td>
</tr>
<tr>
<td align="left">Mutli-line FASTA</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="left">Yes</td>
</tr>
<tr>
<td align="left">Recognize RNA</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="left">Yes</td>
</tr>
<tr>
<td align="left">Read STDIN</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="left">Yes</td>
</tr>
<tr>
<td align="left">Read Gzip</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="left">Yes</td>
</tr>
<tr>
<td align="left">Grep</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="left">Yes</td>
</tr>
<tr>
<td align="left">Mutli-grep</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="left">Yes</td>
</tr>
<tr>
<td align="left">Sample</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="left">Yes</td>
</tr>
<tr>
<td align="left">Subseq</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="left">Yes</td>
</tr>
<tr>
<td align="left">Deduplicate</td>
<td align="center">Yes</td>
<td align="center">Partly</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Partly</td>
<td align="left">--</td>
</tr>
<tr>
<td align="left">Split</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">--</td>
</tr>
<tr>
<td align="left">Barcode split</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">--</td>
</tr>
<tr>
<td align="left">Shuffle</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="left">--</td>
</tr>
<tr>
<td align="left">Locate motifs</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="left">--</td>
</tr>
<tr>
<td align="left">Common</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="left">--</td>
</tr>
<tr>
<td align="left">Clean</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">--</td>
</tr>
<tr>
<td align="left">Transcribe</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="left">Yes</td>
</tr>
<tr>
<td align="left">Translate</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="left">--</td>
</tr>
<tr>
<td align="left">Size select</td>
<td align="center">Indirect</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="left">--</td>
</tr>
<tr>
<td align="left">Rename name</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="left">Yes</td>
</tr>
</tbody>
</table>
<h2 id="datasets">Datasets</h2>
<p>Original datasets included:</p>
<ul>
<li><a href="http://www.arb-silva.de/fileadmin/silva_databases/current/Exports/SILVA_123_SSURef_tax_silva.fasta.gz">SILVA_123_SSURef_tax_silva.fasta.gz</a></li>
<li><a href="ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/H_sapiens/Assembled_chromosomes/seq/">hs_ref_GRCh38.p2_*.mfa.gz</a></li>
</ul>
<p>They are so large, so only subsets are used.</p>
<ol>
<li>
<p><code>dataset_A</code>. Sampling by proption of 0.1 for <code>SILVA_123_SSURef_tax_silva.fasta.gz</code></p>
<pre><code>fakit sample SILVA_123_SSURef_tax_silva.fasta.gz -p 0.1 -o dataset_A.fa.gz
</code></pre>
</li>
<li>
<p><code>dataset_B</code>. Merging chr18,19,20,21,22,Y to a single file</p>
<pre><code>zcat hs_ref_GRCh38.p2_chr{18,19,20,21,22,Y}.mfa.gz | pigz -c &gt; dataset_B.fa.gz
</code></pre>
</li>
</ol>
<p>And some tools do not support RNA sequences,
 and are not able to directly read .gz file,
 so the files are uncompressed, and convert to DNA by <code>fakit seq --rna2dna dataset_A.fa.gz &gt; dataset_A.fa</code>.</p>
<table>
<thead>
<tr>
<th align="center">File</th>
<th align="center">type</th>
<th align="center">num_seqs</th>
<th align="center">min_len</th>
<th align="center">avg_len</th>
<th align="center">max_len</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">dataset_A.fa (261.7M)</td>
<td align="center">DNA</td>
<td align="center">175364</td>
<td align="center">900</td>
<td align="center">1419.6</td>
<td align="center">3725</td>
</tr>
<tr>
<td align="center">dataset_B.fa (346.5M)</td>
<td align="center">DNA</td>
<td align="center">6</td>
<td align="center">46709983</td>
<td align="center">59698489.0</td>
<td align="center">80373285</td>
</tr>
</tbody>
</table>
<h2 id="platform">Platform</h2>
<p>PC:</p>
<ul>
<li>CPU: Intel Core i5-3320M @ 2.60GHz, two cores/4 threads</li>
<li>RAM: DDR3 1600MHz, 12GB</li>
<li>SSD: SAMSUNG 850 EVO 250G, SATA-3</li>
<li>OS: Fedora 23 (Scientific KDE spin),  Kernal: 4.4.3-300.fc23.x86_64</li>
</ul>
<p>Softwares:</p>
<ul>
<li>Perl: perl 5, version 22, subversion 1 (v5.22.1) built for x86_64-linux-thread-multi</li>
<li>Python: Python 2.7.10 (default, Sep  8 2015, 17:20:17) [GCC 5.1.1 20150618 (Red Hat 5.1.1-4)] on linux2</li>
</ul>
<h2 id="test-1-reverse-complement">Test 1. Reverse Complement</h2>
<h3 id="commands">Commands</h3>
<ol>
<li>fakit:   <code>for f in *.fa; do time fakit seq -r -p $f &gt; /dev/null; done</code></li>
<li>fasta_utilities: <code>for f in *.fa; do time reverse_complement.pl $f &gt; /dev/null; done</code></li>
<li>pyfaidx: <code>for f in *.fa; do time faidx -c -r $f &gt; /dev/null; done</code>.
Two runs needed, first run creates fasta index file, and the second one evaluates.</li>
<li>seqmagick: <code>for f in *.fa; do time seqmagick convert --reverse-complement $f - &gt; /dev/null; done</code></li>
<li>seqtk:   <code>for f in *.fa; do time seqtk seq -r $f &gt; /dev/null; done</code></li>
<li><code>revcom_biogo</code> using <a href="https://github.com/biogo/biogo">biogo</a>
(<a href="../files/revcom_biogo.go">source</a>, <a href="../files/revcom_biogo.gz">binary</a> ). (Go).
<code>for f in *.fa; do time ./revcom_biogo $f &gt; /dev/null; done</code></li>
</ol>
<h3 id="results">Results:</h3>
<table>
<thead>
<tr>
<th align="center">Datasets</th>
<th align="center">fakit</th>
<th align="center">fasta_utilities</th>
<th align="center">pyfaidx</th>
<th align="center">seqmagick</th>
<th align="center">seqtk</th>
<th align="center">revcom_biogo</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">dataset_A.fa</td>
<td align="center">5.38s</td>
<td align="center">4.02s</td>
<td align="center">25.98s</td>
<td align="center">13.10s</td>
<td align="center">0.66s</td>
<td align="center">10.92s</td>
</tr>
<tr>
<td align="center">dataset_B.fa</td>
<td align="center">7.68s</td>
<td align="center">2.70s</td>
<td align="center">17.52s</td>
<td align="center">9.42s</td>
<td align="center">0.81s</td>
<td align="center">10.48s</td>
</tr>
</tbody>
</table>
<h2 id="test-2extract-sequencs-by-id-list">Test 2.Extract sequencs by ID list</h2>
<h3 id="id-lists">ID lists</h3>
<p>ID lists come from sampling 80% of dataset_A and shuffling.</p>
<pre><code>$ fakit sample -p 0.8 dataset_A.fa | fakit shuffle | fakit seq -n -i &gt; ids_A.txt
$ wc -l ids_A.txt
140261 ids_A.txt
$ head -n 2 ids_A.txt
GQ103704.1.1352
FR853054.1.1478

$ time fakit sample -p 0.8 dataset_B.fa | fakit shuffle | fakit seq -n -i --id-regexp 'gi\|(.+?)\|'  &gt; ids_B.txt
wc -l ids_B.txt
4 ids_B.txt
$ head -n 2 ids_B.txt
568815574
568815580
</code></pre>
<h3 id="commands_1">Commands</h3>
<ol>
<li>fakit: <code>for g in A B; do time fakit grep -f ids_$g.txt dataset_$g.fa &gt; /dev/null; done</code></li>
<li>fasta_utilities: (using an <a href="https://github.com/shenwei356/fasta_utilities/commit/cac7f14f952fab9bc4a209c6bc2b7cfad47e60d8">fixed</a>
   version of <code>in_list.pl</code>)
   <code>for g in A B; do time in_list.pl -files ids_$g.txt dataset_$g.fa &gt; /dev/null; done</code></li>
<li>fastx_toolkit: can't handle multi-line FASTA files</li>
<li>pyfaidx: unsupported</li>
<li>seqmagick: <code>for g in A B; do time seqmagick convert --include-from-file ids_$g.txt dataset_$g.fa - &gt; /dev/null; done</code></li>
<li>seqtk: <code>for g in A B; do time seqtk subseq  dataset_$g.fa ids_$g.txt &gt; /dev/null; done</code></li>
</ol>
<h3 id="results_1">Results</h3>
<table>
<thead>
<tr>
<th align="left">Datasets</th>
<th align="left">fakit</th>
<th align="left">fasta_utilities</th>
<th align="left">seqmagick</th>
<th align="left">seqtk</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">dataset_A.fa</td>
<td align="left">5.40s</td>
<td align="left">3.88s</td>
<td align="left">8.78s</td>
<td align="left">2.31s</td>
</tr>
<tr>
<td align="left">dataset_A.fa</td>
<td align="left">2.82s</td>
<td align="left">1.50s</td>
<td align="left">4.58s</td>
<td align="left">0.49s</td>
</tr>
</tbody>
</table>
<h2 id="test-3-deduplication">Test 3. Deduplication</h2>
<h3 id="dataset">Dataset</h3>
<p>Randomly extract 10% sequences from dataset_*.fa and merge back and then shuffle.</p>
<pre><code>$ cat &lt;(fakit sample -p 0.1 dataset_A.fa) dataset_A.fa | fakit shuffle &gt; dataset_A_dup.fasta
$ cat &lt;(fakit sample -p 0.2 dataset_B.fa) dataset_B.fa | fakit shuffle &gt; dataset_B_dup.fasta
</code></pre>
<p>Numbers:</p>
<pre><code>dataset_A_dup.fasta: 175364 + 17411 = 192775
dataset_B_dup.fasta: 6 + 2 = 8
</code></pre>
<p>Unique seqs:</p>
<pre><code>$ fakit fa2tab dataset_A_dup.fasta  | cut -f 2 | sort | uniq | wc -l
161864
$ fakit fa2tab dataset_B_dup.fasta  | cut -f 2 | sort | uniq | wc -l
6
</code></pre>
<h3 id="commands_2">Commands</h3>
<p>By sequence</p>
<ol>
<li>fakit: <code>for f in *_dup.fasta; do time fakit rmdup -s $f &gt; /dev/null; done</code></li>
<li>seqmagick: <code>for f in *_dup.fasta; do time seqmagick convert --deduplicate-sequences $f - &gt; /dev/null; done</code></li>
</ol>
<p>By name</p>
<ol>
<li>fakit: <code>for f in *_dup.fasta; do time fakit rmdup -n $f &gt; /dev/null; done</code></li>
<li>fasta_utilities: <code>for f in *_dup.fasta; do time unique_headers.pl $f &gt; /dev/null; done</code></li>
</ol>
<h3 id="results_2">Results</h3>
<p>By sequence</p>
<table>
<thead>
<tr>
<th align="left">Datasets</th>
<th align="left">fakit</th>
<th align="left">seqmagick</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">dataset_A_dup.fasta</td>
<td align="left">5.18s</td>
<td align="left">12.08s</td>
</tr>
<tr>
<td align="left">dataset_B_dup.fasta</td>
<td align="left">5.55s</td>
<td align="left">11.75s</td>
</tr>
</tbody>
</table>
<p>By name</p>
<table>
<thead>
<tr>
<th align="left">Datasets</th>
<th align="left">fakit</th>
<th align="left">fasta_utilities</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">dataset_A_dup.fasta</td>
<td align="left">4.86s</td>
<td align="left">4.35s</td>
</tr>
<tr>
<td align="left">dataset_B_dup.fasta</td>
<td align="left">4.76s</td>
<td align="left">2.44s</td>
</tr>
</tbody>
</table>
<h2 id="test-4-sampling">Test 4. Sampling</h2>
<h3 id="commands_3">Commands</h3>
<p>Sample by number</p>
<ol>
<li>fakit: <code>for g in A B; do time fakit sample -n 1000 dataset_$g.fa &gt; /dev/null; done</code></li>
<li>fasta_utilities: (<em>Not randomly</em>) <code>for g in A B; do time subset_fasta.pl -size 1000 dataset_$g.fa &gt; /dev/null; done</code></li>
<li>fastx_toolkit: unsupported</li>
<li>pyfaidx: unsupported</li>
<li>seqmagick: <code>for g in A B; do time seqmagick convert --sample 1000  dataset_$g.fa - &gt; /dev/null; done</code></li>
<li>seqtk: <code>for g in A B; do time seqtk sample  dataset_$g.fa 1000 &gt; /dev/null; done</code></li>
</ol>
<h3 id="results_3">Results</h3>
<table>
<thead>
<tr>
<th align="left">Datasets</th>
<th align="left">fakit</th>
<th align="left">fasta_utilities</th>
<th align="left">seqmagick</th>
<th align="left">seqtk</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">dataset_A.fa</td>
<td align="left">2.68s</td>
<td align="left">3.01s</td>
<td align="left">5.75s</td>
<td align="left">0.38s</td>
</tr>
<tr>
<td align="left">dataset_A.fa</td>
<td align="left">3.62s</td>
<td align="left">1.96s</td>
<td align="left">7.81s</td>
<td align="left">0.60s</td>
</tr>
</tbody>
</table>
<h2 id="test-5-spliting">Test 5. Spliting</h2>
<p>TODO</p></div>
            
        </div>

        <footer class="col-md-12">
            <hr>
            
            <p>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>

        <script src="../js/jquery-1.10.2.min.js"></script>
        <script src="../js/bootstrap-3.0.3.min.js"></script>
        <script src="../js/highlight.pack.js"></script>
        <script>var base_url = '..';</script>
        <script data-main="../mkdocs/js/search.js" src="../mkdocs/js/require.js"></script>
        <script src="../js/base.js"></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
                        <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                    </div>
                    <div class="modal-body">
                        <p>
                            From here you can search these documents. Enter
                            your search terms below.
                        </p>
                        <form role="form">
                            <div class="form-group">
                                <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                            </div>
                        </form>
                        <div id="mkdocs-search-results"></div>
                    </div>
                    <div class="modal-footer">
                    </div>
                </div>
            </div>
        </div>

    </body>
</html>
