<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        <meta name="author" content="shenwei356">
        
        <link rel="shortcut icon" href="../img/favicon.ico">

	<title>Benchmark - fakit - FASTA/Q kit</title>

        <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="../css/font-awesome-4.5.0.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.css">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

            ga('create', 'UA-28948626-3', 'blog.shenwei.me/fakit');
            ga('send', 'pageview');
        </script>
        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="..">fakit - FASTA/Q kit</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="../download/">Download</a>
                    </li>
                
                
                
                    <li >
                        <a href="../usage/">Usage</a>
                    </li>
                
                
                
                    <li >
                        <a href="../tutorial/">Tutorial</a>
                    </li>
                
                
                
                    <li class="active">
                        <a href="./">Benchmark</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li >
                        <a rel="next" href="../tutorial/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li class="disabled">
                        <a rel="prev" >
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/shenwei356/fakit">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            
                <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#benchmark">Benchmark</a></li>
        
            <li><a href="#softwares">Softwares</a></li>
        
            <li><a href="#features">Features</a></li>
        
            <li><a href="#datasets">Datasets</a></li>
        
            <li><a href="#platform">Platform</a></li>
        
            <li><a href="#tests">Tests</a></li>
        
            <li><a href="#results">Results</a></li>
        
    
    </ul>
</div></div>
                <div class="col-md-9" role="main">

<h1 id="benchmark">Benchmark</h1>
<h2 id="softwares">Softwares</h2>
<ol>
<li><a href="https://github.com/shenwei356/fakit">fakit</a>. (Go).
   Version <a href="https://github.com/shenwei356/fakit/releases/tag/v0.2.7">v0.2.7</a>.</li>
<li><a href="https://github.com/jimhester/fasta_utilities">fasta_utilities</a>. (Perl).
   Version <a href="https://github.com/jimhester/fasta_utilities/tree/3dcc0bc6bf1e97839476221c26984b1789482579">3dcc0bc</a>.
   Lots of dependencies to install.</li>
<li><a href="http://hannonlab.cshl.edu/fastx_toolkit/">fastx_toolkit</a>. (Perl).
   Version <a href="http://hannonlab.cshl.edu/fastx_toolkit/fastx_toolkit_0.0.13_binaries_Linux_2.6_amd64.tar.bz2">0.0.13</a>.
   Can't handle multi-line FASTA files.</li>
<li><a href="http://seqmagick.readthedocs.io/en/latest/index.html#installation">seqmagick</a>. (Python).
   Version 0.6.1</li>
<li><a href="https://github.com/lh3/seqtk">seqtk</a>. (C).
   Version <a href="https://github.com/lh3/seqtk/tree/fb85aad4ce1fc7b3d4543623418a1ae88fe1cea6">1.1-r92-dirty</a>.</li>
</ol>
<p>Not used:</p>
<ol>
<li><a href="https://github.com/mdshw5/pyfaidx">pyfaidx</a>. (Python).
   Version <a href="https://pypi.python.org/packages/source/p/pyfaidx/pyfaidx-0.4.7.1.tar.gz#md5=f33604a3550c2fa115ac7d33b952127d">0.4.7.1</a>. <em>Not used, because it exhausted my memory (10G) when computing reverse-complement on a 5GB fasta file of 250 bp.</em></li>
</ol>
<p>A Python script <a href="https://github.com/shenwei356/memusg">memusg</a> was used
to compute running time and peak memory usage of a process.</p>
<h2 id="features">Features</h2>
<table>
<thead>
<tr>
<th align="left">Features</th>
<th>fakit</th>
<th>fasta_utilities</th>
<th>fastx_toolkit</th>
<th>pyfaidx</th>
<th>seqmagick</th>
<th>seqtk</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Cross-platform</td>
<td>Yes</td>
<td>Partly</td>
<td>Partly</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Mutli-line FASTA</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Read FASTQ</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Mutli-line FASTQ</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Validate bases</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Recognize RNA</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Read STDIN</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Read gzip</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Write gzip</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>--</td>
</tr>
<tr>
<td align="left">Search by motifs</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>--</td>
</tr>
<tr>
<td align="left">Sample seqs</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Subseq</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Deduplicate seqs</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>Partly</td>
<td>--</td>
</tr>
<tr>
<td align="left">Split seqs</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>Partly</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Split by seq</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Shuffle seqs</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Sort seqs</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>--</td>
</tr>
<tr>
<td align="left">Locate motifs</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Common seqs</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Clean bases</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Transcribe</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Translate</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
</tr>
<tr>
<td align="left">Size select</td>
<td>Indirect</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
</tr>
<tr>
<td align="left">Rename head</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
</tbody>
</table>
<h2 id="datasets">Datasets</h2>
<p>All test data is available here: <a href="http://bioinf.shenwei.me/fakit-benchmark-data.tar.gz">fakit-benchmark-data.tar.gz</a>  (1.7G)</p>
<h3 id="dataset_afa-large-number-of-short-sequences">dataset_A.fa - large number of short sequences</h3>
<p>Dataset A is reference genomes DNA sequences of gastrointestinal tract from
<a href="http://hmpdacc.org/">NIH Human Microbiome Project</a>:
<a href="http://downloads.hmpdacc.org/data/reference_genomes/body_sites/Gastrointestinal_tract.nuc.fsa"><code>Gastrointestinal_tract.nuc.fsa</code></a> (FASTA format, ~2.7G).</p>
<h3 id="dataset_bfa-small-number-of-large-sequences">dataset_B.fa - small number of large sequences</h3>
<p>Dataset B is Human genome from <a href="http://uswest.ensembl.org/info/data/ftp/index.html">ensembl</a>.</p>
<ul>
<li>Genome DNA:  <a href="ftp://ftp.ensembl.org/pub/release-84/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz"><code>Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz</code></a> (Gzipped FASTA file, ~900M)</li>
<li>GTF file:  <a href="ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.gtf.gz"><code>Homo_sapiens.GRCh38.84.gtf.gz</code></a> (~44M)</li>
<li>BED file: <code>Homo_sapiens.GRCh38.84.bed.gz</code> was converted from <code>Homo_sapiens.GRCh38.84.gtf.gz</code> by  <a href="http://bedops.readthedocs.org/en/latest/content/reference/file-management/conversion/gtf2bed.html?highlight=gtf2bed"><code>gtf2bed</code></a>  with command<pre><code>$ zcat Homo_sapiens.GRCh38.84.gtf.gz | gtf2bed --do-not-sort | gzip -c &gt; Homo_sapiens.GRCh38.84.bed.gz
</code></pre>
</li>
</ul>
<p>Summary</p>
<pre><code>$ fakit stat *.fa
file           seq_format   seq_type   num_seqs   min_len        avg_len       max_len
dataset_A.fa   FASTA        DNA          67,748        56       41,442.5     5,976,145
dataset_B.fa   FASTA        DNA             194       970   15,978,096.5   248,956,422
</code></pre>
<h3 id="sequence-id-list">Sequence ID list</h3>
<p>Parts of sequences IDs was sampled and shuffled from original data.
They were used in test of extracting sequences by ID list.</p>
<p>Commands:</p>
<pre><code>$ fakit sample -p 0.3 dataset_A.fa | fakit seq --name --only-id | shuf &gt; ids_A.txt
$ fakit sample -p 0.3 dataset_B.fa | fakit seq --name --only-id | shuf &gt; ids_B.txt
</code></pre>
<p>Numbers:</p>
<pre><code>$ wc -l ids_*
20138 ids_A.txt
   58 ids_B.txt
</code></pre>
<h3 id="bed-file">BED file</h3>
<p>Only BED data of chromosome 19 was used in test of subsequence with BED file:</p>
<pre><code>$ zcat Homo_sapiens.GRCh38.84.bed.gz | grep -E "^19" | gzip -c &gt; chr19.bed.gz
</code></pre>
<h2 id="platform">Platform</h2>
<p>PC:</p>
<ul>
<li>CPU: Intel Core i5-3320M @ 2.60GHz, two cores/4 threads</li>
<li>RAM: DDR3 1600MHz, 12GB</li>
<li>SSD: SAMSUNG 850 EVO 250G, SATA-3</li>
<li>OS: Fedora 24 (Scientific KDE spin),  Kernal: 4.6.3-200.fc24.x86_64</li>
</ul>
<p>Softwares:</p>
<ul>
<li>Perl: perl 5, version 22, subversion 1 (v5.22.2) built for x86_64-linux-thread-multi</li>
<li>Python: Python 2.7.11 (default, Jul 10 2016, 20:58:20) [GCC 6.1.1 20160621 (Red Hat 6.1.1-3)] on linux2</li>
</ul>
<h2 id="tests">Tests</h2>
<p>Automatic benchmark and plotting scripts are available at:  <a href="https://github.com/shenwei356/fakit/tree/master/benchmark">https://github.com/shenwei356/fakit/tree/master/benchmark</a>.</p>
<p>All tests were repeated 3 times ( ~20 min for one time),
and average time and peak memory ware used for plotting.</p>
<p>All data were readed once before tests began to minimize the influence of page cache.</p>
<h3 id="test-1-reverse-complement">Test 1. Reverse Complement</h3>
<p>Output sequences of all Softwares were not wrapped to fixed length.</p>
<p><code>revcom_biogo</code> (<a href="https://github.com/shenwei356/fakit/blob/master/benchmark/revcom_biogo.go">source</a>,
 <a href="https://github.com/shenwei356/fakit/blob/master/benchmark/revcom_biogo?raw=true">binary</a> ),
 a tool written in Golang using <a href="https://github.com/biogo/biogo">biogo</a> package,
 was also used for comparison of FASTA file parsing performance.</p>
<p><a href="https://github.com/shenwei356/fakit/blob/master/benchmark/run_benchmark_01_revcom.sh">Commands</a></p>
<h3 id="test-2-extracting-sequences-by-id-list">Test 2. Extracting sequences by ID list</h3>
<p><a href="https://github.com/shenwei356/fakit/blob/master/benchmark/run_benchmark_02_exctact_by_id_list.sh">Commands</a></p>
<h3 id="test-3-sampling-by-number">Test 3. Sampling by number</h3>
<p>Note that different softwares have different sampling strategies,
the peak memory may depends on size of sampled sequences.</p>
<p><a href="https://github.com/shenwei356/fakit/blob/master/benchmark/run_benchmark_03_sampling.sh">Commands</a></p>
<h3 id="test-4-removing-duplicates-by-sequence-content">Test 4. Removing duplicates by sequence content</h3>
<p><a href="https://github.com/shenwei356/fakit/blob/master/benchmark/run_benchmark_04_remove_duplicated_seqs_by_seq.sh">Commands</a></p>
<h3 id="test-5-subsequence-with-bed-file">Test 5. Subsequence with BED file</h3>
<p><a href="https://github.com/shenwei356/fakit/blob/master/benchmark/run_benchmark_05_subseq_with_bed.sh">Commands</a></p>
<h2 id="results">Results</h2>
<p>fakit version: v0.2.7</p>
<p><img alt="benchmark-5tests.csv.png" src="../benchmark/benchmark.5tests.csv.png" /></p>
<h3 id="test-of-multiple-threads">Test of multiple threads:</h3>
<p>From the results, 2 threads/CPU is enough, so the default threads of fakit is 2.</p>
<p><img alt="benchmark-5tests.csv.png" src="../benchmark/fakit_multi_threads/benchmark.5tests.csv.png" /></p>
<h3 id="tests-on-different-file-sizes">Tests on different file sizes</h3>
<p>Files are generated by replicating Human genome chr1 for N times.</p>
<p><img alt="benchmark.fakit.files_size.csv.png" src="../benchmark/fakit_file_size/benchmark.fakit.files_size.csv.png" /></p>
<div id="disqus_thread"></div>

<script>
/**
* RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
* LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables
*/
/*
var disqus_config = function () {
this.page.url = PAGE_URL; // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');

s.src = '//fastakit.disqus.com/embed.js';

s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>

<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript" rel="nofollow">comments powered by Disqus.</a></noscript></div>
            
        </div>

        <footer class="col-md-12">
            <hr>
            
            <p>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>

        <script src="../js/jquery-1.10.2.min.js"></script>
        <script src="../js/bootstrap-3.0.3.min.js"></script>
        <script src="../js/highlight.pack.js"></script>
        <script>var base_url = '..';</script>
        <script data-main="../mkdocs/js/search.js" src="../mkdocs/js/require.js"></script>
        <script src="../js/base.js"></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
                        <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                    </div>
                    <div class="modal-body">
                        <p>
                            From here you can search these documents. Enter
                            your search terms below.
                        </p>
                        <form role="form">
                            <div class="form-group">
                                <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                            </div>
                        </form>
                        <div id="mkdocs-search-results"></div>
                    </div>
                    <div class="modal-footer">
                    </div>
                </div>
            </div>
        </div>

    </body>
</html>
