<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        <meta name="author" content="shenwei356">
        
        <link rel="shortcut icon" href="../img/favicon.ico">

	<title>Tutorial - fakit - FASTA kit</title>

        <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="../css/font-awesome-4.5.0.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.css">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

            ga('create', 'UA-28948626-3', 'blog.shenwei.me/fakit');
            ga('send', 'pageview');
        </script>
        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="..">fakit - FASTA kit</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="../download/">Download</a>
                    </li>
                
                
                
                    <li >
                        <a href="../usage/">Usage</a>
                    </li>
                
                
                
                    <li class="active">
                        <a href="./">Tutorial</a>
                    </li>
                
                
                
                    <li >
                        <a href="../benchmark/">Benchmark</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li >
                        <a rel="next" href="../usage/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../benchmark/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/shenwei356/fakit">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            
                <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#tutorial">Tutorial</a></li>
        
            <li><a href="#play-with-mirna-hairpins">Play with miRNA hairpins</a></li>
        
            <li><a href="#bacteria-genome">Bacteria genome</a></li>
        
    
    </ul>
</div></div>
                <div class="col-md-9" role="main">

<h1 id="tutorial">Tutorial</h1>
<h3 id="play-with-mirna-hairpins">Play with miRNA hairpins</h3>
<h4 id="dataset">Dataset</h4>
<p><a href="ftp://mirbase.org/pub/mirbase/21/hairpin.fa.gz"><code>hairpin.fa.gz</code></a>
from <a href="ftp://mirbase.org/pub/mirbase/21/">The miRBase Sequence Database -- Release 21</a></p>
<h4 id="quick-glance">Quick glance</h4>
<ol>
<li>
<p>Amount</p>
<pre><code>$ fakit stat hairpin.fa.gz
file             seq_type    num_seqs    min_len    avg_len    max_len
hairpin.fa.gz         RNA      28,645         39        103      2,354
</code></pre>
</li>
<li>
<p>First 10 bases</p>
<pre><code>$ zcat hairpin.fa.gz | fakit subseq -r 1:10 | fakit sort -s  | fakit seq -t rna  -s | head -n 10
AAAAAAAAAA
AAAAAAAAAA
AAAAAAAAAG
AAAAAAAAAG
AAAAAAAAAG
AAAAAAAAAU
AAAAAAAAGG
AAAAAAACAU
AAAAAAACGA
AAAAAAAUUA
</code></pre>
<p>hmm, nothing special.</p>
</li>
</ol>
<h4 id="repeated-hairpin-sequences">Repeated hairpin sequences.</h4>
<p>The result shows the most conserved miRNAs among different species,
mir-29b, mir-125, mir-19b-1 and mir-20a.
And the dre-miR-430c has the most multicopies in <em>Danio rerio</em>.</p>
<pre><code>$ fakit rmdup -s -i hairpin.fa.gz -o clean.fa.gz -d duplicated.fa.gz -D duplicated.detail.txt

$ head -n 5 duplicated.detail.txt
18      dre-mir-430c-1, dre-mir-430c-2, dre-mir-430c-3, dre-mir-430c-4, dre-mir-430c-5, dre-mir-430c-6, dre-mir-430c-7, dre-mir-430c-8, dre-mir-430c-9, dre-mir-430c-10, dre-mir-430c-11, dre-mir-430c-12, dre-mir-430c-13, dre-mir-430c-14, dre-mir-430c-15, dre-mir-430c-16, dre-mir-430c-17, dre-mir-430c-18
16      hsa-mir-29b-2, mmu-mir-29b-2, rno-mir-29b-2, ptr-mir-29b-2, ggo-mir-29b-2, ppy-mir-29b-2, sla-mir-29b, mne-mir-29b, ppa-mir-29b-2, bta-mir-29b-2, mml-mir-29b-2, eca-mir-29b-2, aja-mir-29b, oar-mir-29b-1, oar-mir-29b-2, rno-mir-29b-3
15      dme-mir-125, dps-mir-125, dan-mir-125, der-mir-125, dgr-mir-125-1, dgr-mir-125-2, dmo-mir-125, dpe-mir-125-2, dpe-mir-125-1, dpe-mir-125-3, dse-mir-125, dsi-mir-125, dvi-mir-125, dwi-mir-125, dya-mir-125
13      hsa-mir-19b-1, ggo-mir-19b-1, age-mir-19b-1, ppa-mir-19b-1, ppy-mir-19b-1, ptr-mir-19b-1, mml-mir-19b-1, sla-mir-19b-1, lla-mir-19b-1, mne-mir-19b-1, bta-mir-19b, oar-mir-19b, chi-mir-19b
13      hsa-mir-20a, ssc-mir-20a, ggo-mir-20a, age-mir-20, ppa-mir-20, ppy-mir-20a, ptr-mir-20a, mml-mir-20a, sla-mir-20, lla-mir-20, mne-mir-20, bta-mir-20a, eca-mir-20a
</code></pre>
<h4 id="hairpins-in-different-species">Hairpins in different species</h4>
<ol>
<li>
<p>Split sequences by species</p>
<pre><code>$ fakit split hairpin.fa.gz -i --id-regexp "^([\w]+)\-"
</code></pre>
</li>
<li>
<p>Species with most miRNA hairpins</p>
<pre><code>$ fakit stat hairpin.id_*.gz | sort -k3,3nr
hairpin.id_hsa.fa.gz           RNA       1,881         41       81.9        180
hairpin.id_mmu.fa.gz           RNA       1,193         39       83.4        147
hairpin.id_bta.fa.gz           RNA         808         53       80.1        149
hairpin.id_gga.fa.gz           RNA         740         48       91.5        169
hairpin.id_eca.fa.gz           RNA         715         52      104.6        145
</code></pre>
</li>
</ol>
<p>For human miRNA hairpins</p>
<ol>
<li>
<p>Length distribution (
 <a href="https://github.com/shenwei356/bio_scripts/blob/master/plot/plot_distribution.py">plot_distribution.py</a> )</p>
<pre><code>$ fakit fa2tab hairpin.id_hsa.fa.gz -n -i -l | cut -f 3  | plot_distribution.py -o hairpin.id_hsa.fa.gz.lendist.png
</code></pre>
<p><img alt="hairpin.id_hsa.fa.gz.lendist.png" src="../files/hairpin/hairpin.id_hsa.fa.gz.lendist.png" /></p>
</li>
</ol>
<h3 id="bacteria-genome">Bacteria genome</h3>
<h4 id="dataset_1">Dataset</h4>
<p><a href="http://www.ncbi.nlm.nih.gov/nuccore/110645304">Pseudomonas aeruginosa PAO1</a>,
files:</p>
<ul>
<li>Genbank file <a href="../files/PAO1/PAO1.gb"><code>PAO1.gb</code></a></li>
<li>Genome FASTA file <a href="../files/PAO1/PAO1.fasta"><code>PAO1.fasta</code></a></li>
<li>GTF file <a href="../files/PAO1/PAO1.gtf"><code>PAO1.gtf</code></a> was created with <a href="https://github.com/shenwei356/bio_scripts/blob/master/file_formats/extract_features_from_genbank_file.py"><code>extract_features_from_genbank_file.py</code></a>, by<pre><code>extract_features_from_genbank_file.py  PAO1.gb -t . -f gtf &gt; PAO1.gtf
</code></pre>
</li>
</ul>
<h4 id="motif-distribution">Motif distribution</h4>
<p>Motifs</p>
<pre><code>$ cat motifs.fa
&gt;GTAGCGS
GTAGCGS
&gt;GGWGKTCG
GGWGKTCG
</code></pre>
<ol>
<li>
<p>sliding</p>
<pre><code>$ fakit sliding --id-ncbi --circle-genome --step 20000 --window 200000 PAO1.fasta -o PAO1.fasta.sliding.fa

$ fakit stat PAO1.fasta.sliding.fa
file                     seq_type    num_seqs    min_len    avg_len    max_len
PAO1.fasta.sliding.fa         DNA         314    200,000    200,000    200,000
</code></pre>
</li>
<li>
<p>locating motifs</p>
<pre><code>$ fakit locate --id-ncbi -i -d -f motifs.fa  PAO1.fasta.sliding.fa -o  PAO1.fasta.sliding.fa.motifs.tsv
</code></pre>
</li>
<li>
<p>ploting distribution (<a href="../files/PAO1/plot_motif_distribution.R">plot_motif_distribution.R</a>)</p>
<pre><code># preproccess
$ perl -ne 'if (/_sliding:(\d+)-(\d+)\t(.+)/) {$loc= $1 + 100000; print "$loc\t$3\n";} else {print}' PAO1.fasta.sliding.fa.motifs.tsv  &gt; PAO1.fasta.sliding.fa.motifs.tsv2

# plot
$ ./plot_motif_distribution.R
</code></pre>
<p>Result</p>
<p><img alt="motif_distribution.png" src="../files/PAO1/motif_distribution.png" /></p>
</li>
</ol>
<h4 id="find-multicopy-genes">Find multicopy genes</h4>
<ol>
<li>
<p>Get all CDS sequences</p>
<pre><code>$ fakit subseq --id-ncbi --gtf PAO1.gtf --feature cds PAO1.fasta -o PAO1.cds.fasta

$ fakit stat *.fasta
file              seq_type    num_seqs      min_len      avg_len      max_len
PAO1.cds.fasta         DNA       5,572           72      1,003.8       16,884
PAO1.fasta             DNA           1    6,264,404    6,264,404    6,264,404
</code></pre>
</li>
<li>
<p>Get 1000 bp upstream sequence of CDS</p>
<pre><code>$ fakit subseq --id-ncbi --gtf PAO1.gtf --feature cds PAO1.fasta --up-stream 1000 --only-flank -o PAO1.cds.u1k.fasta
</code></pre>
</li>
<li>
<p>Get duplicated sequences</p>
<pre><code>$ fakit rmdup -s -i PAO1.cds.fasta -o PAO1.cds.uniq.fasta -d PAO1.cds.dup.fasta -D PAO1.cds.dup.text

$ cat PAO1.cds.dup.text
6       NC_002516.2_500104:501120:-, NC_002516.2_2556948:2557964:+, NC_002516.2_3043750:3044766:-, NC_002516.2_3842274:3843290:-, NC_002516.2_4473623:4474639:+, NC_002516.2_5382796:5383812:-
2       NC_002516.2_2073555:2075438:+, NC_002516.2_4716660:4718543:+
2       NC_002516.2_2072935:2073558:+, NC_002516.2_4716040:4716663:+
2       NC_002516.2_2075452:2076288:+, NC_002516.2_4718557:4719393:+
</code></pre>
</li>
</ol></div>
            
        </div>

        <footer class="col-md-12">
            <hr>
            
            <p>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>

        <script src="../js/jquery-1.10.2.min.js"></script>
        <script src="../js/bootstrap-3.0.3.min.js"></script>
        <script src="../js/highlight.pack.js"></script>
        <script>var base_url = '..';</script>
        <script data-main="../mkdocs/js/search.js" src="../mkdocs/js/require.js"></script>
        <script src="../js/base.js"></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
                        <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                    </div>
                    <div class="modal-body">
                        <p>
                            From here you can search these documents. Enter
                            your search terms below.
                        </p>
                        <form role="form">
                            <div class="form-group">
                                <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                            </div>
                        </form>
                        <div id="mkdocs-search-results"></div>
                    </div>
                    <div class="modal-footer">
                    </div>
                </div>
            </div>
        </div>

    </body>
</html>
