<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <meta name="description" content="fakit - Practical FASTA kit">
        <meta name="author" content="shenwei356">
        
        <link rel="shortcut icon" href="./img/favicon.ico">

	<title>fakit - FASTA kit</title>

        <link href="./css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="./css/font-awesome-4.5.0.css" rel="stylesheet">
        <link href="./css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="./css/highlight.css">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

            ga('create', 'UA-28948626-3', 'blog.shenwei.me/fakit');
            ga('send', 'pageview');
        </script>
        
    </head>

    <body class="homepage">

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href=".">fakit - FASTA kit</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li class="active">
                        <a href=".">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="download/">Download</a>
                    </li>
                
                
                
                    <li >
                        <a href="usage/">Usage</a>
                    </li>
                
                
                
                    <li >
                        <a href="tutorial/">Tutorial</a>
                    </li>
                
                
                
                    <li >
                        <a href="benchmark/">Benchmark</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li class="disabled">
                        <a rel="next" >
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="download/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/shenwei356/fakit">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            
                <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#fakit-swiss-army-knife-of-fasta-format">fakit - Swiss army knife of FASTA format</a></li>
        
            <li><a href="#introduction">Introduction</a></li>
        
            <li><a href="#features">Features</a></li>
        
            <li><a href="#subcommands">Subcommands</a></li>
        
            <li><a href="#usage-examples">Usage &amp;&amp; Examples</a></li>
        
            <li><a href="#benchmark">Benchmark</a></li>
        
            <li><a href="#contact">Contact</a></li>
        
            <li><a href="#license">License</a></li>
        
    
    </ul>
</div></div>
                <div class="col-md-9" role="main">

<h1 id="fakit-swiss-army-knife-of-fasta-format">fakit - Swiss army knife of FASTA format</h1>
<p>Documents  : <a href="http://shenwei356.github.io/fakit">http://shenwei356.github.io/fakit</a></p>
<p>Source code: <a href="https://github.com/shenwei356/fakit">https://github.com/shenwei356/fakit</a></p>
<h2 id="introduction">Introduction</h2>
<p>FASTA is a basic format for storing nucleotide and protein sequences.
The manipulation of FASTA file includes converting, clipping, searching, filtering,
deduplication, splitting, shuffling, sampling and so on.
Existed tools only implemented parts of the functions,
and some of them are only available for specific operating systems.
Furthermore, the complicated installation process of dependencies packages and
running environment also make them less friendly to common users.</p>
<p>Fakit is a cross-platform, efficient, and practical FASTA manipulations tool
that is friendly for researchers to complete wide ranges of FASTA file processing.
The suite supports plain or gzip-compressed input and output
from either standard stream or files, therefore, it could be easily used in pipelines.</p>
<h2 id="features">Features</h2>
<ul>
<li><strong>Cross-platform</strong> (Linux/Windows/Mac OS X/OpenBSD/FreeBSD,
  see <a href="http://shenwei356.github.io/fakit/download/">download</a>)</li>
<li><strong>Light weight and out-of-the-box, no dependencies, no compilation, no configuration</strong>
  (see <a href="http://shenwei356.github.io/fakit/download/">download</a>)</li>
<li><strong>Fast</strong> (see <a href="./#performance-comparison-with-other-tools">benchmark</a>),
  <strong>multiple-threads supported</strong> (see <a href="./#speedup-with-multi-threads">benchmark</a>).</li>
<li><strong>Practical functions supported by 14 subcommands</strong> (see subcommands and
  <a href="http://shenwei356.github.io/fakit/usage/">usage</a> )</li>
<li><strong>Well documented</strong> (detailed <a href="http://shenwei356.github.io/fakit/usage/">usage</a>
  and <a href="http://shenwei356.github.io/fakit/benchmark/">benchmark</a> )</li>
<li><strong>Support STDIN and gziped input/output file, easy being used in pipe</strong></li>
<li>Support custom sequence ID regular expression (especially useful for quering with ID list)</li>
<li>Reproducible results (configurable rand seed in <code>sample</code> and <code>shuffle</code>)</li>
<li>Well organized source code, friendly to use and easy to extend.</li>
</ul>
<p>Features comparison</p>
<table>
<thead>
<tr>
<th align="left">Features</th>
<th>fakit</th>
<th>fasta_utilities</th>
<th>fastx_toolkit</th>
<th>pyfaidx</th>
<th>seqmagick</th>
<th>seqtk</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Cross-platform</td>
<td>Yes</td>
<td>Partly</td>
<td>Partly</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Mutli-line FASTA</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Validate bases</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Recognize RNA</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Read STDIN</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Read gzip</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Write gzip</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>--</td>
</tr>
<tr>
<td align="left">Search by motifs</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Sample seqs</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Subseq</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Deduplicate seqs</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>Partly</td>
<td>--</td>
</tr>
<tr>
<td align="left">Split seqs</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>Partly</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Split by seq</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Shuffle seqs</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Sort seqs</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>--</td>
</tr>
<tr>
<td align="left">Locate motifs</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Common seqs</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Clean bases</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
</tr>
<tr>
<td align="left">Transcribe</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td align="left">Translate</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
</tr>
<tr>
<td align="left">Size select</td>
<td>Indirect</td>
<td>Yes</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
<td>--</td>
</tr>
<tr>
<td align="left">Rename head</td>
<td>--</td>
<td>Yes</td>
<td>--</td>
<td>--</td>
<td>Yes</td>
<td>Yes</td>
</tr>
</tbody>
</table>
<h2 id="subcommands">Subcommands</h2>
<p>Sequence and subsequence</p>
<ul>
<li><code>seq</code>        transform sequences (revserse, complement, extract ID...)</li>
<li><code>subseq</code>     get subsequences by region/gtf/bed, including flanking sequences</li>
<li><code>sliding</code>    sliding sequences, circle genome supported</li>
<li><code>stat</code>       simple statistics of FASTA files</li>
</ul>
<p>Format conversion</p>
<ul>
<li><code>fa2tab</code>     covert FASTA to tabular format (and length/GC content/GC skew) to filter and sort</li>
<li><code>tab2fa</code>     covert tabular format to FASTA format</li>
</ul>
<p>Searching</p>
<ul>
<li><code>grep</code>       search sequences by pattern(s) of name or sequence motifs</li>
<li><code>locate</code>     locate subsequences/motifs</li>
</ul>
<p>Set operations</p>
<ul>
<li><code>rmdup</code>      remove duplicated sequences by id/name/sequence</li>
<li><code>common</code>     find common sequences of multiple files by id/name/sequence</li>
<li><code>split</code>      split sequences into files by id/seq region/size/parts</li>
<li><code>sample</code>     sample sequences by number or proportion</li>
</ul>
<p>Ordering</p>
<ul>
<li><code>shuffle</code>    shuffle sequences</li>
<li><code>sort</code>       sort sequences by id/name/sequence</li>
</ul>
<p>Global Flags</p>
<pre><code>      --alphabet-guess-seq-length int   length of sequence prefix of the first FASTA record based on which fakit guesses the sequence type (default 10000)
  -c, --chunk-size int                  chunk size (attention: unit is FASTA records not lines) (default 1000)
      --id-ncbi                         FASTA head is NCBI-style, e.g. &gt;gi|110645304|ref|NC_002516.2| Pseud...
      --id-regexp string                regular expression for parsing ID (default &quot;^([^\\s]+)\\s?&quot;)
  -w, --line-width int                  line width (0 for no wrap) (default 60)
  -o, --out-file string                 out file (&quot;-&quot; for stdout, suffix .gz for gzipped out) (default &quot;-&quot;)
      --quiet                           be quiet and do not show extra information
  -t, --seq-type string                 sequence type (dna|rna|protein|unlimit|auto) (for auto, it automatically detect by the first sequence) (default &quot;auto&quot;)
  -j, --threads int                     number of CPUs. (default value depends on your device) (default 4)

</code></pre>

<h2 id="usage-examples">Usage &amp;&amp; Examples</h2>
<p><a href="http://shenwei356.github.io/fakit/usage/">Usage and examples</a></p>
<p><a href="http://shenwei356.github.io/fakit/tutorial/">Tutorial</a></p>
<h2 id="benchmark">Benchmark</h2>
<p>Details: <a href="http://shenwei356.github.io/fakit/benchmark/">http://shenwei356.github.io/fakit/benchmark/</a></p>
<h3 id="performance-comparison-with-other-tools">Performance comparison with other tools</h3>
<p>All tests were repeated 4 times</p>
<p><img alt="benchmark_colorful.png" src="./benchmark/benchmark_colorful.png" /></p>
<h3 id="speedup-with-multi-threads">Speedup with multi-threads</h3>
<p><img alt="benchmark_colorful.png" src="./benchmark/fakit_multi_threads/benchmark_colorful.png" /></p>
<h2 id="contact">Contact</h2>
<p>Email me for any problem when using fakit. shenwei356(at)gmail.com</p>
<p><a href="https://github.com/shenwei356/fakit/issues">Create an issue</a> to report bugs,
propose new functions or ask for help.</p>
<h2 id="license">License</h2>
<p><a href="https://github.com/shenwei356/bio_scripts/blob/master/LICENSE">MIT License</a></p></div>
            
        </div>

        <footer class="col-md-12">
            <hr>
            
            <p>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>

        <script src="./js/jquery-1.10.2.min.js"></script>
        <script src="./js/bootstrap-3.0.3.min.js"></script>
        <script src="./js/highlight.pack.js"></script>
        <script>var base_url = '.';</script>
        <script data-main="./mkdocs/js/search.js" src="./mkdocs/js/require.js"></script>
        <script src="./js/base.js"></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
                        <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                    </div>
                    <div class="modal-body">
                        <p>
                            From here you can search these documents. Enter
                            your search terms below.
                        </p>
                        <form role="form">
                            <div class="form-group">
                                <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                            </div>
                        </form>
                        <div id="mkdocs-search-results"></div>
                    </div>
                    <div class="modal-footer">
                    </div>
                </div>
            </div>
        </div>

    </body>
</html>

<!--
MkDocs version : 0.15.3
Build Date UTC : 2016-03-28 02:59:52.406232
-->
