



<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="SeqKit - a cross-platform and ultrafast toolkit for FASTA/Q file manipulation">
      
      
      
        <meta name="author" content="Wei Shen">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="files/favicon.ico">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-4.4.0">
    
    
      
        <title>SeqKit - Ultrafast FASTA/Q kit</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/application.0284f74d.css">
      
        <link rel="stylesheet" href="assets/stylesheets/application-palette.01803549.css">
      
      
        
        
        <meta name="theme-color" content="#009688">
      
    
    
      <script src="assets/javascripts/modernizr.74668098.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="assets/fonts/material-icons.css">
    
      <link rel="manifest" href="manifest.webmanifest">
    
    
    
      
        
<script>
  window.ga = window.ga || function() {
    (ga.q = ga.q || []).push(arguments)
  }
  ga.l = +new Date
  /* Setup integration and send page view */
  ga("create", "UA-28948626-9", "bioinf.shenwei.me/seqkit")
  ga("set", "anonymizeIp", true)
  ga("send", "pageview")
  /* Register handler to log search on blur */
  document.addEventListener("DOMContentLoaded", () => {
    if (document.forms.search) {
      var query = document.forms.search.query
      query.addEventListener("blur", function() {
        if (this.value) {
          var path = document.location.pathname;
          ga("send", "pageview", path + "?q=" + this.value)
        }
      })
    }
  })
</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
      
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="teal" data-md-color-accent="blue-grey">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#seqkit-a-cross-platform-and-ultrafast-toolkit-for-fastaq-file-manipulation" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="." title="SeqKit - Ultrafast FASTA/Q kit" class="md-header-nav__button md-logo">
          
            <i class="md-icon">school</i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            <span class="md-header-nav__topic">
              SeqKit - Ultrafast FASTA/Q kit
            </span>
            <span class="md-header-nav__topic">
              
                Home
              
            </span>
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
          
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  

<a href="https://github.com/shenwei356/seqkit/" title="Go to repository" class="md-source" data-md-source="github">
  
    <div class="md-source__icon">
      <svg viewBox="0 0 24 24" width="24" height="24">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    GitHub
  </div>
</a>
          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="." title="SeqKit - Ultrafast FASTA/Q kit" class="md-nav__button md-logo">
      
        <i class="md-icon">school</i>
      
    </a>
    SeqKit - Ultrafast FASTA/Q kit
  </label>
  
    <div class="md-nav__source">
      


  

<a href="https://github.com/shenwei356/seqkit/" title="Go to repository" class="md-source" data-md-source="github">
  
    <div class="md-source__icon">
      <svg viewBox="0 0 24 24" width="24" height="24">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Home
      </label>
    
    <a href="." title="Home" class="md-nav__link md-nav__link--active">
      Home
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" title="Introduction" class="md-nav__link">
    Introduction
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" title="Table of Contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" title="Features" class="md-nav__link">
    Features
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#features-comparison" title="Features comparison" class="md-nav__link">
    Features comparison
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subcommands" title="Subcommands" class="md-nav__link">
    Subcommands
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" title="Installation" class="md-nav__link">
    Installation
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#method-1-download-binaries-latest-stabledev-version" title="Method 1: Download binaries (latest stable/dev version)" class="md-nav__link">
    Method 1: Download binaries (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-2-install-via-conda-latest-stable-version" title="Method 2: Install via conda  (latest stable version)" class="md-nav__link">
    Method 2: Install via conda  (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-3-install-via-homebrew-latest-stable-version" title="Method 3: Install via homebrew  (latest stable version)" class="md-nav__link">
    Method 3: Install via homebrew  (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-4-for-go-developer-latest-stabledev-version" title="Method 4: For Go developer (latest stable/dev version)" class="md-nav__link">
    Method 4: For Go developer (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-5-docker-based-installation-latest-stabledev-version" title="Method 5: Docker based installation (latest stable/dev version)" class="md-nav__link">
    Method 5: Docker based installation (latest stable/dev version)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bash-completion" title="Bash-completion" class="md-nav__link">
    Bash-completion
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#technical-details-and-guides-for-use" title="Technical details and guides for use" class="md-nav__link">
    Technical details and guides for use
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastaq-format-parsing" title="FASTA/Q format parsing" class="md-nav__link">
    FASTA/Q format parsing
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence-formats-and-types" title="Sequence formats and types" class="md-nav__link">
    Sequence formats and types
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence-id" title="Sequence ID" class="md-nav__link">
    Sequence ID
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fasta-index" title="FASTA index" class="md-nav__link">
    FASTA index
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parallelization-of-cpu-intensive-jobs" title="Parallelization of CPU intensive jobs" class="md-nav__link">
    Parallelization of CPU intensive jobs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#memory-occupation" title="Memory occupation" class="md-nav__link">
    Memory occupation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reproducibility" title="Reproducibility" class="md-nav__link">
    Reproducibility
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage-examples" title="Usage &amp;&amp; Examples" class="md-nav__link">
    Usage &amp;&amp; Examples
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#benchmark" title="Benchmark" class="md-nav__link">
    Benchmark
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citation" title="Citation" class="md-nav__link">
    Citation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#acknowledgements" title="Acknowledgements" class="md-nav__link">
    Acknowledgements
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contact" title="Contact" class="md-nav__link">
    Contact
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#license" title="License" class="md-nav__link">
    License
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="download/" title="Download" class="md-nav__link">
      Download
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3">
    
    <label class="md-nav__link" for="nav-3">
      Documents
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-3">
        Documents
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="usage/" title="Usage" class="md-nav__link">
      Usage
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="faq/" title="FAQ" class="md-nav__link">
      FAQ
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="tutorial/" title="Tutorial" class="md-nav__link">
      Tutorial
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="benchmark/" title="Benchmark" class="md-nav__link">
      Benchmark
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="note/" title="Notes" class="md-nav__link">
      Notes
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="https://github.com/shenwei356" title="More tools" class="md-nav__link">
      More tools
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" title="Introduction" class="md-nav__link">
    Introduction
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" title="Table of Contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" title="Features" class="md-nav__link">
    Features
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#features-comparison" title="Features comparison" class="md-nav__link">
    Features comparison
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subcommands" title="Subcommands" class="md-nav__link">
    Subcommands
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" title="Installation" class="md-nav__link">
    Installation
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#method-1-download-binaries-latest-stabledev-version" title="Method 1: Download binaries (latest stable/dev version)" class="md-nav__link">
    Method 1: Download binaries (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-2-install-via-conda-latest-stable-version" title="Method 2: Install via conda  (latest stable version)" class="md-nav__link">
    Method 2: Install via conda  (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-3-install-via-homebrew-latest-stable-version" title="Method 3: Install via homebrew  (latest stable version)" class="md-nav__link">
    Method 3: Install via homebrew  (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-4-for-go-developer-latest-stabledev-version" title="Method 4: For Go developer (latest stable/dev version)" class="md-nav__link">
    Method 4: For Go developer (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-5-docker-based-installation-latest-stabledev-version" title="Method 5: Docker based installation (latest stable/dev version)" class="md-nav__link">
    Method 5: Docker based installation (latest stable/dev version)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bash-completion" title="Bash-completion" class="md-nav__link">
    Bash-completion
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#technical-details-and-guides-for-use" title="Technical details and guides for use" class="md-nav__link">
    Technical details and guides for use
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastaq-format-parsing" title="FASTA/Q format parsing" class="md-nav__link">
    FASTA/Q format parsing
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence-formats-and-types" title="Sequence formats and types" class="md-nav__link">
    Sequence formats and types
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequence-id" title="Sequence ID" class="md-nav__link">
    Sequence ID
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fasta-index" title="FASTA index" class="md-nav__link">
    FASTA index
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parallelization-of-cpu-intensive-jobs" title="Parallelization of CPU intensive jobs" class="md-nav__link">
    Parallelization of CPU intensive jobs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#memory-occupation" title="Memory occupation" class="md-nav__link">
    Memory occupation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reproducibility" title="Reproducibility" class="md-nav__link">
    Reproducibility
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage-examples" title="Usage &amp;&amp; Examples" class="md-nav__link">
    Usage &amp;&amp; Examples
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#benchmark" title="Benchmark" class="md-nav__link">
    Benchmark
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citation" title="Citation" class="md-nav__link">
    Citation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#acknowledgements" title="Acknowledgements" class="md-nav__link">
    Acknowledgements
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contact" title="Contact" class="md-nav__link">
    Contact
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#license" title="License" class="md-nav__link">
    License
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/shenwei356/seqkit/edit/master/docs/index.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <h1 id="seqkit-a-cross-platform-and-ultrafast-toolkit-for-fastaq-file-manipulation">SeqKit - a cross-platform and ultrafast toolkit for FASTA/Q file manipulation</h1>
<ul>
<li><strong>Documents:</strong> <a href="http://bioinf.shenwei.me/seqkit">http://bioinf.shenwei.me/seqkit</a>
(<a href="http://bioinf.shenwei.me/seqkit/usage/"><strong>Usage</strong></a>,
<a href="http://bioinf.shenwei.me/seqkit/faq/"><strong>FAQ</strong></a>,
<a href="http://bioinf.shenwei.me/seqkit/tutorial/"><strong>Tutorial</strong></a>,
and 
<a href="http://bioinf.shenwei.me/seqkit/benchmark/"><strong>Benchmark</strong></a>)</li>
<li><strong>Source code:</strong> <a href="https://github.com/shenwei356/seqkit">https://github.com/shenwei356/seqkit</a>
<a href="https://github.com/shenwei356/seqkit"><img alt="GitHub stars" src="https://img.shields.io/github/stars/shenwei356/seqkit.svg?style=social&amp;label=Star&amp;?maxAge=2592000" /></a>
<a href="https://github.com/shenwei356/seqkit/blob/master/LICENSE"><img alt="license" src="https://img.shields.io/github/license/shenwei356/seqkit.svg?maxAge=2592000" /></a>
<a href="https://travis-ci.org/shenwei356/seqkit"><img alt="Build Status" src="https://travis-ci.org/shenwei356/seqkit.svg?branch=master" /></a></li>
<li><strong>Latest version:</strong> <a href="https://github.com/shenwei356/seqkit/releases"><img alt="Latest Version" src="https://img.shields.io/github/release/shenwei356/seqkit.svg?style=flat?maxAge=86400" /></a>
<a href="http://bioinf.shenwei.me/seqkit/download/"><img alt="Github Releases" src="https://img.shields.io/github/downloads/shenwei356/seqkit/latest/total.svg?maxAge=3600" /></a>
<a href="http://bioinf.shenwei.me/seqkit/download/"><img alt="Cross-platform" src="https://img.shields.io/badge/platform-any-ec2eb4.svg?style=flat" /></a>
<a href="https://anaconda.org/bioconda/seqkit"><img alt="Anaconda Cloud" src="https://anaconda.org/bioconda/seqkit/badges/version.svg" /></a></li>
<li><strong><a href="#citation">Please cite</a>:</strong> <a href="https://doi.org/10.1371/journal.pone.0163962"><img alt="doi" src="https://img.shields.io/badge/doi-10.1371%2Fjournal.pone.0163962-blue.svg?style=flat" /></a></li>
</ul>
<h2 id="introduction">Introduction</h2>
<p>FASTA and FASTQ are basic and ubiquitous formats for storing nucleotide and
protein sequences. Common manipulations of FASTA/Q file include converting,
searching, filtering, deduplication, splitting, shuffling, and sampling.
Existing tools only implement some of these manipulations,
and not particularly efficiently, and some are only available for certain
operating systems. Furthermore, the complicated installation process of
required packages and running environments can render these programs less
user friendly.</p>
<p>This project describes a cross-platform ultrafast comprehensive
toolkit for FASTA/Q processing. SeqKit provides executable binary files for
all major operating systems, including Windows, Linux, and Mac OS X, and can
be directly used without any dependencies or pre-configurations.
SeqKit demonstrates competitive performance in execution time and memory
usage compared to similar tools. The efficiency and usability of SeqKit
enable researchers to rapidly accomplish common FASTA/Q file manipulations.</p>
<h2 id="table-of-contents">Table of Contents</h2>
<!-- START doctoc generated TOC please keep comment here to allow auto update -->

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

<ul>
<li><a href="#features">Features</a></li>
<li><a href="#subcommands">Subcommands</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#bash-completion">Bash-completion</a></li>
<li><a href="#technical-details-and-guides-for-use">Technical details and guides for use</a></li>
<li><a href="#usage--examples">Usage &amp;&amp; Examples</a></li>
<li><a href="#benchmark">Benchmark</a></li>
<li><a href="#citation">Citation</a></li>
<li><a href="#acknowledgements">Acknowledgements</a></li>
<li><a href="#contact">Contact</a></li>
<li><a href="#license">License</a></li>
</ul>
<!-- END doctoc generated TOC please keep comment here to allow auto update -->

<h2 id="features">Features</h2>
<ul>
<li><strong>Cross-platform</strong> (Linux/Windows/Mac OS X/OpenBSD/FreeBSD,
  see <a href="http://bioinf.shenwei.me/seqkit/download/">download</a>)</li>
<li><strong>Light weight and out-of-the-box, no dependencies, no compilation, no configuration</strong>
  (see <a href="http://bioinf.shenwei.me/seqkit/download/">download</a>)</li>
<li><strong>UltraFast</strong> (see <a href="#benchmark">benchmark</a>),
  <strong>multiple-CPUs supported</strong></li>
<li><strong>Practical functions supported by 27 subcommands</strong> (see subcommands and
  <a href="http://bioinf.shenwei.me/seqkit/usage/">usage</a> )</li>
<li><strong>Supporting <a href="#bash-completion">Bash-completion</a></strong></li>
<li><strong>Well documented</strong> (detailed <a href="http://bioinf.shenwei.me/seqkit/usage/">usage</a>
  and <a href="http://bioinf.shenwei.me/seqkit/benchmark/">benchmark</a> )</li>
<li><strong>Seamlessly parsing both FASTA and FASTQ formats</strong></li>
<li><strong>Supporting STDIN and gzipped input/output file, easy being used in pipe</strong>
  , <strong>writing gzip file is very fast (10X of <code>gzip</code>, 4X of <code>pigz</code>)</strong> by
  using package <a href="https://github.com/klauspost/pgzip">pgzip</a></li>
<li><strong>Supporting custom sequence ID regular expression</strong> (especially useful for searching with ID list)</li>
<li>Reproducible results (configurable rand seed in <code>sample</code> and <code>shuffle</code>)</li>
<li>Well organized source code, friendly to use and easy to extend</li>
</ul>
<h3 id="features-comparison">Features comparison</h3>
<table>
<thead>
<tr>
<th align="left">Categories</th>
<th align="left">Features</th>
<th align="center">seqkit</th>
<th align="center">fasta_utilities</th>
<th align="center">fastx_toolkit</th>
<th align="center">pyfaidx</th>
<th align="center">seqmagick</th>
<th align="center">seqtk</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><strong>Formats support</strong></td>
<td align="left">Multi-line FASTA</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">FASTQ</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Multi-line  FASTQ</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Validating sequences</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Supporting RNA</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr>
<td align="left"><strong>Functions</strong></td>
<td align="left">Searching by motifs</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Sampling</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Extracting sub-sequence</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Removing duplicates</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Partly</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Splitting</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Partly</td>
<td align="center">--</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Splitting by seq</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Shuffling</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Sorting</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Locating motifs</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Common sequences</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Cleaning bases</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Transcription</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Translation</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Filtering by size</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Renaming header</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr>
<td align="left"><strong>Other features</strong></td>
<td align="left">Cross-platform</td>
<td align="center">Yes</td>
<td align="center">Partly</td>
<td align="center">Partly</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Reading STDIN</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Reading gzipped file</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr>
<td align="left"></td>
<td align="left">Writing gzip file</td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">--</td>
<td align="center">Yes</td>
<td align="center">--</td>
</tr>
</tbody>
</table>
<p><strong>Note 1</strong>: See <a href="http://bioinf.shenwei.me/seqkit/benchmark/#softwares">version information</a> of the softwares.</p>
<p><strong>Note 2</strong>: See <a href="http://bioinf.shenwei.me/seqkit/usage/">usage</a> for detailed options of seqkit.</p>
<h2 id="subcommands">Subcommands</h2>
<p>27 functional subcommands in total.</p>
<p><strong>Sequence and subsequence</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#seq"><code>seq</code></a>          transform sequences (revserse, complement, extract ID...)</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#subseq"><code>subseq</code></a>    get subsequences by region/gtf/bed, including flanking sequences</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#sliding"><code>sliding</code></a>  sliding sequences, circular genome supported</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#stats"><code>stats</code></a>      simple statistics of FASTA/Q files</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#faidx"><code>faidx</code></a>      create FASTA index file and extract subsequence</li>
</ul>
<p><strong>Format conversion</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#fx2tab"><code>fx2tab</code></a>        convert FASTA/Q to tabular format (and length/GC content/GC skew)</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#tab2fx"><code>tab2fx</code></a>        convert tabular format to FASTA/Q format</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#fq2fa"><code>fq2fa</code></a>          convert FASTQ to FASTA</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#convert"><code>convert</code></a>      convert FASTQ quality encoding between Sanger, Solexa and Illumina</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#translate"><code>translate</code></a>  translate DNA/RNA to protein sequence (supporting ambiguous bases)</li>
</ul>
<p><strong>Searching</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#grep"><code>grep</code></a>        search sequences by ID/name/sequence/sequence motifs, mismatch allowed</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#locate"><code>locate</code></a>    locate subsequences/motifs</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#amplicon"><code>amplicon</code></a> retrieve amplicon (or specific region around it) via primer(s)</li>
</ul>
<p><strong>Set operations</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#head"><code>head</code></a>            print first N FASTA/Q records</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#range"><code>range</code></a>          print FASTA/Q records in a range (start:end)</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#sample"><code>sample</code></a>        sample sequences by number or proportion</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#rmdup"><code>rmdup</code></a>          remove duplicated sequences by id/name/sequence</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#duplicate"><code>duplicate</code></a>  duplicate sequences N times</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#common"><code>common</code></a>        find common sequences of multiple files by id/name/sequence</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#split"><code>split</code></a>          split sequences into files by id/seq region/size/parts (mainly for FASTA)</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#split2"><code>split2</code></a>        split sequences into files by size/parts (FASTA, PE/SE FASTQ)</li>
</ul>
<p><strong>Edit</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#replace"><code>replace</code></a>  replace name/sequence by regular expression</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#rename"><code>rename</code></a>    rename duplicated IDs</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#restart"><code>restart</code></a>  reset start position for circular genome</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#concat"><code>concat</code></a>    concatenate sequences with same ID from multiple files</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#mutate"><code>mutate</code></a>    edit sequence (point mutation, insertion, deletion)</li>
</ul>
<p><strong>Ordering</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#shuffle"><code>shuffle</code></a>  shuffle sequences</li>
<li><a href="https://bioinf.shenwei.me/seqkit/usage/#sort"><code>sort</code></a>        sort sequences by id/name/sequence</li>
</ul>
<p><strong>Misc</strong></p>
<ul>
<li><code>version</code>   print version information and check for update</li>
<li><code>genautocomplete</code> generate shell autocompletion script</li>
</ul>
<h2 id="installation">Installation</h2>
<p>Go to <a href="http://bioinf.shenwei.me/seqkit/download">Download Page</a> for more download options and changelogs.</p>
<p><code>SeqKit</code> is implemented in <a href="https://golang.org/">Go</a> programming language,
 executable binary files <strong>for most popular operating systems</strong> are freely available
  in <a href="https://github.com/shenwei356/seqkit/releases">release</a> page.</p>
<h4 id="method-1-download-binaries-latest-stabledev-version">Method 1: Download binaries (latest stable/dev version)</h4>
<p>Just <a href="https://github.com/shenwei356/seqkit/releases">download</a> compressed
executable file of your operating system,
and decompress it with <code>tar -zxvf *.tar.gz</code> command or other tools.
And then:</p>
<ol>
<li>
<p><strong>For Linux-like systems</strong></p>
<ol>
<li>
<p>If you have root privilege simply copy it to <code>/usr/local/bin</code>:</p>
<pre><code>sudo cp seqkit /usr/local/bin/
</code></pre>
</li>
<li>
<p>Or copy to anywhere in the environment variable <code>PATH</code>:</p>
<pre><code>mkdir -p $HOME/bin/; cp seqkit $HOME/bin/
</code></pre>
</li>
</ol>
</li>
<li>
<p><strong>For windows</strong>, just copy <code>seqkit.exe</code> to <code>C:\WINDOWS\system32</code>.</p>
</li>
</ol>
<h4 id="method-2-install-via-conda-latest-stable-version">Method 2: Install via conda  (latest stable version) <a href="https://anaconda.org/bioconda/seqkit"><img alt="Anaconda Cloud" src="https://anaconda.org/bioconda/seqkit/badges/version.svg" /></a> <a href="https://anaconda.org/bioconda/seqkit"><img alt="downloads" src="https://anaconda.org/bioconda/seqkit/badges/downloads.svg" /></a></h4>
<pre><code>conda install -c bioconda seqkit
</code></pre>
<h4 id="method-3-install-via-homebrew-latest-stable-version">Method 3: Install via homebrew  (latest stable version)</h4>
<pre><code>brew install brewsci/bio/seqkit
</code></pre>
<h4 id="method-4-for-go-developer-latest-stabledev-version">Method 4: For Go developer (latest stable/dev version)</h4>
<pre><code>go get -u github.com/shenwei356/seqkit/seqkit
</code></pre>
<h4 id="method-5-docker-based-installation-latest-stabledev-version">Method 5: Docker based installation (latest stable/dev version)</h4>
<p><a href="https://docs.docker.com/engine/installation/#supported-platforms">Install Docker</a></p>
<p>git clone this repo:</p>
<pre><code>git clone https://github.com/shenwei356/seqkit
</code></pre>
<p>Run the following commands:</p>
<pre><code>cd seqkit
docker build -t shenwei356/seqkit .
docker run -it shenwei356/seqkit:latest
</code></pre>
<h2 id="bash-completion">Bash-completion</h2>
<p>Note: The current version supports Bash only.
This should work for *nix systems with Bash installed.</p>
<p>Howto:</p>
<ol>
<li>
<p>run: <code>seqkit genautocomplete</code></p>
</li>
<li>
<p>create and edit <code>~/.bash_completion</code> file if you don't have it.</p>
<pre><code>nano ~/.bash_completion
</code></pre>
<p>add the following:</p>
<pre><code>for bcfile in ~/.bash_completion.d/* ; do
  . $bcfile
done
</code></pre>
</li>
</ol>
<h2 id="technical-details-and-guides-for-use">Technical details and guides for use</h2>
<h3 id="fastaq-format-parsing">FASTA/Q format parsing</h3>
<p>SeqKit uses author's lightweight and high-performance bioinformatics packages
<a href="https://github.com/shenwei356/bio">bio</a> for FASTA/Q parsing,
which has <a href="https://github.com/shenwei356/bio#fastaq-parsing">high performance</a>
close to the
famous C lib <a href="https://github.com/attractivechaos/klib/">klib</a> (<a href="https://github.com/attractivechaos/klib/blob/master/kseq.h">kseq.h</a>).</p>
<p><img alt="" src="https://github.com/shenwei356/bio/raw/master/benchmark/benchmark.tsv.png" /></p>
<p><s>Seqkit calls <code>pigz</code> (much faster than <code>gzip</code>) or <code>gzip</code> to decompress .gz file if they are available.
So please <strong>install <a href="http://zlib.net/pigz/">pigz</a> to gain better parsing performance for gzipped data</strong>.</s>
Seqkit does not call <code>pigz</code> or <code>gzip</code> any more since v0.8.1,
Because it does not always increase the speed.
But you can still utilize <code>pigz</code> or <code>gzip</code> by <code>pigz -d -c seqs.fq.gz | seqkit xxx</code>.</p>
<p>Seqkit uses package <a href="https://github.com/klauspost/pgzip">pgzip</a> to write gzip file,
which is very fast (<strong>10X of <code>gzip</code>, 4X of <code>pigz</code></strong>) and the gzip file would be slighty larger.</p>
<h3 id="sequence-formats-and-types">Sequence formats and types</h3>
<p>SeqKit seamlessly support FASTA and FASTQ format.
Sequence format is automatically detected.
All subcommands except for <code>faidx</code> can handle both formats.
And only when some commands (<code>subseq</code>, <code>split</code>, <code>sort</code> and <code>shuffle</code>)
which utilise FASTA index to improve perfrmance for large files in two pass mode
(by flag <code>--two-pass</code>), only FASTA format is supported.</p>
<p>Sequence type (DNA/RNA/Protein) is automatically detected by leading subsequences
of the first sequences in file or STDIN. The length of the leading subsequences
is configurable by global flag <code>--alphabet-guess-seq-length</code> with default value
of 10000. If length of the sequences is less than that, whole sequences will
be checked.</p>
<h3 id="sequence-id">Sequence ID</h3>
<p>By default, most softwares, including <code>seqkit</code>, take the leading non-space
letters as sequence identifier (ID). For example,</p>
<table>
<thead>
<tr>
<th align="left">FASTA header</th>
<th align="left">ID</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">&gt;123456 gene name</td>
<td align="left">123456</td>
</tr>
<tr>
<td align="left">&gt;longname</td>
<td align="left">longname</td>
</tr>
<tr>
<td align="left">&gt;gi&#124;110645304&#124;ref&#124;NC_002516.2&#124; Pseudomona</td>
<td align="left">gi&#124;110645304&#124;ref&#124;NC_002516.2&#124;</td>
</tr>
</tbody>
</table>
<p>But for some sequences from NCBI,
e.g. <code>&gt;gi|110645304|ref|NC_002516.2| Pseudomona</code>, the ID is <code>NC_002516.2</code>.
In this case, we could set sequence ID parsing regular expression by global flag
<code>--id-regexp "\|([^\|]+)\| "</code> or just use flag <code>--id-ncbi</code>. If you want
the <code>gi</code> number, then use <code>--id-regexp "^gi\|([^\|]+)\|"</code>.</p>
<h3 id="fasta-index">FASTA index</h3>
<p>For some commands, including <code>subseq</code>, <code>split</code>, <code>sort</code> and <code>shuffle</code>,
when input files are (plain or gzipped) FASTA files,
FASTA index would be optional used for
rapid access of sequences and reducing memory occupation.</p>
<p>ATTENTION: the <code>.seqkit.fai</code> file created by SeqKit is slightly different from <code>.fai</code> file
created by <code>samtools</code>. SeqKit uses full sequence head instead of just ID as key.</p>
<h3 id="parallelization-of-cpu-intensive-jobs">Parallelization of CPU intensive jobs</h3>
<p>The validation of sequences bases and complement process of sequences
are parallelized for large sequences.</p>
<p>Parsing of line-based files, including BED/GFF file and ID list file are also parallelized.</p>
<p>The Parallelization is implemented by multiple goroutines in golang
 which are similar to but much
lighter weight than threads. The concurrency number is configurable with global
flag <code>-j</code> or <code>--threads</code> (default value: 1 for single-CPU PC, 2 for others).</p>
<h3 id="memory-occupation">Memory occupation</h3>
<p>Most of the subcommands do not read whole FASTA/Q records in to memory,
including <code>stat</code>, <code>fq2fa</code>, <code>fx2tab</code>, <code>tab2fx</code>, <code>grep</code>, <code>locate</code>, <code>replace</code>,
 <code>seq</code>, <code>sliding</code>, <code>subseq</code>.</p>
<p>Note that when using <code>subseq --gtf | --bed</code>, if the GTF/BED files are too
big, the memory usage will increase.
You could use <code>--chr</code> to specify chromesomes and <code>--feature</code> to limit features.</p>
<p>Some subcommands need to store sequences or heads in memory, but there are
strategy to reduce memory occupation, including <code>rmdup</code> and <code>common</code>.
When comparing with sequences, MD5 digest could be used to replace sequence by
flag <code>-m</code> (<code>--md5</code>).</p>
<p>Some subcommands could either read all records or read the files twice by flag
<code>-2</code> (<code>--two-pass</code>), including <code>sample</code>, <code>split</code>, <code>shuffle</code> and <code>sort</code>.
They use FASTA index for rapid acccess of sequences and reducing memory occupation.</p>
<h3 id="reproducibility">Reproducibility</h3>
<p>Subcommands <code>sample</code> and <code>shuffle</code> use random function, random seed could be
given by flag <code>-s</code> (<code>--rand-seed</code>). This makes sure that sampling result could be
reproduced in different environments with same random seed.</p>
<h2 id="usage-examples">Usage &amp;&amp; Examples</h2>
<p><a href="http://bioinf.shenwei.me/seqkit/usage/">Usage and examples</a></p>
<p><a href="http://bioinf.shenwei.me/seqkit/tutorial/">Tutorial</a></p>
<h2 id="benchmark">Benchmark</h2>
<p>More details: <a href="http://bioinf.shenwei.me/seqkit/benchmark/">http://bioinf.shenwei.me/seqkit/benchmark/</a></p>
<p>Datasets:</p>
<pre><code>$ seqkit stat *.fa
file          format  type   num_seqs        sum_len  min_len       avg_len      max_len
dataset_A.fa  FASTA   DNA      67,748  2,807,643,808       56      41,442.5    5,976,145
dataset_B.fa  FASTA   DNA         194  3,099,750,718      970  15,978,096.5  248,956,422
dataset_C.fq  FASTQ   DNA   9,186,045    918,604,500      100           100          100
</code></pre>
<p>SeqKit version: v0.3.1.1</p>
<p>FASTA:</p>
<p><img alt="benchmark-5tests.tsv.png" src="benchmark/benchmark.5tests.tsv.png" /></p>
<p>FASTQ:</p>
<p><img alt="benchmark-5tests.tsv.png" src="benchmark/benchmark.5tests.tsv.C.png" /></p>
<h2 id="citation">Citation</h2>
<p><strong>W Shen</strong>, S Le, Y Li*, F Hu*. SeqKit: a cross-platform and ultrafast toolkit for FASTA/Q file manipulation.
<strong><em>PLOS ONE</em></strong>. <a href="https://doi.org/10.1371/journal.pone.0163962">doi:10.1371/journal.pone.0163962</a>.</p>
<h2 id="acknowledgements">Acknowledgements</h2>
<p>We thank <a href="https://github.com/jameslz">Lei Zhang</a> for testing of SeqKit,
and also thank <a href="https://github.com/jimhester/">Jim Hester</a>,
author of <a href="https://github.com/jimhester/fasta_utilities">fasta_utilities</a>,
for advice on early performance improvements of for FASTA parsing
and <a href="https://twitter.com/BBToolsBio">Brian Bushnell</a>,
author of <a href="https://sourceforge.net/projects/bbmap/">BBMaps</a>,
for advice on naming SeqKit and adding accuracy evaluation in benchmarks.
We also thank Nicholas C. Wu from the Scripps Research Institute,
USA for commenting on the manuscript
and <a href="http://guangchuangyu.github.io/">Guangchuang Yu</a>
from State Key Laboratory of Emerging Infectious Diseases,
The University of Hong Kong, HK for advice on the manuscript.</p>
<p>We thank <a href="https://github.com/penglbio">Li Peng</a> for reporting many bugs.</p>
<h2 id="contact">Contact</h2>
<p>Email me for any problem when using seqkit. shenwei356(at)gmail.com</p>
<p><a href="https://github.com/shenwei356/seqkit/issues">Create an issue</a> to report bugs,
propose new functions or ask for help.</p>
<h2 id="license">License</h2>
<p><a href="https://github.com/shenwei356/seqkit/blob/master/LICENSE">MIT License</a></p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
        
          <a href="download/" title="Download" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Download
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="assets/javascripts/application.245445c6.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:"."}})</script>
      
    
  </body>
</html>