<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        <meta name="author" content="shenwei356">
        
        <link rel="shortcut icon" href="../img/favicon.ico">

	<title>Usage - fakit - FASTA kit</title>

        <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="../css/font-awesome-4.5.0.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.css">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

            ga('create', 'UA-28948626-3', 'blog.shenwei.me/fakit');
            ga('send', 'pageview');
        </script>
        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="..">fakit - FASTA kit</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="../download/">Download</a>
                    </li>
                
                
                
                    <li class="active">
                        <a href="./">Usage</a>
                    </li>
                
                
                
                    <li >
                        <a href="../tutorial/">Tutorial</a>
                    </li>
                
                
                
                    <li >
                        <a href="../benchmark/">Benchmark</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li >
                        <a rel="next" href="../download/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../tutorial/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/shenwei356/fakit">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            
                <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#usage-and-examples">Usage and Examples</a></li>
        
            <li><a href="#technical-details-and-guides-for-use">Technical details and guides for use</a></li>
        
            <li><a href="#fakit">fakit</a></li>
        
            <li><a href="#seq">seq</a></li>
        
            <li><a href="#subseq">subseq</a></li>
        
            <li><a href="#sliding">sliding</a></li>
        
            <li><a href="#stat">stat</a></li>
        
            <li><a href="#fq2fa">fq2fa</a></li>
        
            <li><a href="#fx2tab-tab2fx">fx2tab &amp; tab2fx</a></li>
        
            <li><a href="#grep">grep</a></li>
        
            <li><a href="#locate">locate</a></li>
        
            <li><a href="#rmdup">rmdup</a></li>
        
            <li><a href="#common">common</a></li>
        
            <li><a href="#split">split</a></li>
        
            <li><a href="#sample">sample</a></li>
        
            <li><a href="#replace">replace</a></li>
        
            <li><a href="#shuffle">shuffle</a></li>
        
            <li><a href="#sort">sort</a></li>
        
    
    </ul>
</div></div>
                <div class="col-md-9" role="main">

<h1 id="usage-and-examples">Usage and Examples</h1>
<h2 id="technical-details-and-guides-for-use">Technical details and guides for use</h2>
<h3 id="reading-fastaq">Reading FASTA/Q</h3>
<p>fakit use author's bioinformatics packages <a href="https://github.com/shenwei356/bio">bio</a>
for FASTA/Q parsing, which <strong>asynchronously parse FASTA/Q records and buffer them
in chunks</strong>. The parser return one chunk of records for each call.</p>
<p><strong>Asynchronous parsing</strong> saves much time because these's no waiting interval for
parsed records being handled.
The strategy of <strong>records chunks</strong> reduces data exchange in parallelly handling
of sequences, which could also improve performance.</p>
<p>Since using of buffers and chunks, the memory occupation will be higher than
cases of reading sequence one by one.
The default value of chunk size (configurable by global flag <code>-c</code> or <code>--chunk-size</code>)
is 1000, which is suitable for manipulating "small" sequences, e.g. FASTQ.
But for big genomes like human genome, smaller chunk size is prefered, e.g. 1.
And the buffer size is configurable by  global flag <code>-b</code> or <code>--buffer-size</code>
(default value is the number of CPUs), therefore, you may set with smaller
value to reduce memory usage.</p>
<p><strong><em>In summary, set smaller value for <code>-c</code> and <code>-b</code> when handling big FASTA file
like human genomes.</em></strong></p>
<h3 id="parallelization-of-cpu-intensive-jobs">Parallelization of CPU intensive jobs</h3>
<p>Most of the manipulations of FASTA/Q files are I/O intensive, to improve the
performance, asynchronous parsing strategy is used.</p>
<p>For CPU intensive jobs like <code>grep</code> with regular expressions, <code>locate</code> with
sequence motifs, and <code>subseq</code> by GTF/BED files. The processes are parallelized
with MapReduce model by multiple goroutine in golang, similar to but much
lighter weight than threads. The concurrency number is configurable with global
flag <code>-j</code> or <code>--threads</code>.</p>
<p>Most of the time you can just use the default value. i.e. the number of CPUs
of your computer.</p>
<h3 id="memory-occupation">Memory occupation</h3>
<p>Most of the subcommands do not read whole FASTA/Q records in to memory,
including <code>stat</code>, <code>fq2fa</code>, <code>fx2tab</code>, <code>tab2fx</code>, <code>grep</code>, <code>locate</code>, <code>replace</code>,
 <code>seq</code>, <code>sliding</code>, <code>subseq</code>. The just temporarily buffer chunks of records.</p>
<p>Some subcommands need to store sequences or heads in memory, but there are
strategy to reduce memory occupation, including <code>rmdup</code> and <code>common</code>.
When comparing with sequences, MD5 digest could be used to replace sequence by
flag <code>-m</code> (<code>--md5</code>).</p>
<p>Some subcommands could either read all records or read the files twice by flag
<code>-2</code> (<code>--two-pass</code>), including <code>sample</code> and <code>split</code>.</p>
<p>Two subcommands must read all records in memory right now, <code>shuffle</code> and <code>sort</code>.
But I'll improve this later.</p>
<h2 id="fakit">fakit</h2>
<p>Usage</p>
<pre><code>fakit --  a cross-platform and efficient suit for FASTA/Q file manipulation

Version: 0.1.7

Author: Wei Shen &lt;shenwei356@gmail.com&gt;

Documents  : http://shenwei356.github.io/fakit
Source code: https://github.com/shenwei356/fakit

Usage:
  fakit [command]

Available Commands:
  common      find common sequences of multiple files by id/name/sequence
  fq2fa       covert FASTQ to FASTA
  fx2tab      covert FASTA/Q to tabular format (with length/GC content/GC skew) to filter and sort
  grep        search sequences by pattern(s) of name or sequence motifs
  locate      locate subsequences/motifs
  replace     replace name/sequence/by regular expression
  rmdup       remove duplicated sequences by id/name/sequence
  sample      sample sequences by number or proportion
  seq         transform sequences (revserse, complement, extract ID...)
  shuffle     shuffle sequences
  sliding     sliding sequences, circle genome supported
  sort        sort sequences by id/name/sequence/length
  split       split sequences into files by id/seq region/size/parts
  stat        simple statistics of FASTA files
  subseq      get subsequences by region/gtf/bed, including flanking sequences
  tab2fx      covert tabular format to FASTA/Q format

Flags:
      --alphabet-guess-seq-length int   length of sequence prefix of the first FASTA record based on which fakit guesses the sequence type (default 10000)
  -b, --buffer-size int                 buffer size of chunks (default value is the CPUs number of your computer) (default 4)
  -c, --chunk-size int                  chunk size (attention: unit is FASTA records not lines) (default 1000)
      --id-ncbi                         FASTA head is NCBI-style, e.g. &gt;gi|110645304|ref|NC_002516.2| Pseud...
      --id-regexp string                regular expression for parsing ID (default &quot;^([^\\s]+)\\s?&quot;)
  -w, --line-width int                  line width when outputing FASTA format (0 for no wrap) (default 60)
  -o, --out-file string                 out file (&quot;-&quot; for stdout, suffix .gz for gzipped out) (default &quot;-&quot;)
      --quiet                           be quiet and do not show extra information
  -t, --seq-type string                 sequence type (dna|rna|protein|unlimit|auto) (for auto, it automatically detect by the first sequence) (default &quot;auto&quot;)
  -j, --threads int                     number of CPUs. (default value is the CPUs number of your computer) (default 4)

Use &quot;fakit [command] --help&quot; for more information about a command.

</code></pre>

<h3 id="datasets">Datasets</h3>
<p>Datasets from <a href="ftp://mirbase.org/pub/mirbase/21/">The miRBase Sequence Database -- Release 21</a></p>
<ul>
<li><a href="ftp://mirbase.org/pub/mirbase/21/hairpin.fa.gz"><code>hairpin.fa.gz</code></a></li>
<li><a href="ftp://mirbase.org/pub/mirbase/21/mature.fa.gz"><code>mature.fa.gz</code></a></li>
<li><a href="ftp://mirbase.org/pub/mirbase/21/miRNA.diff.gz"><code>miRNA.diff.gz</code></a></li>
</ul>
<p>Human genome from <a href="http://uswest.ensembl.org/info/data/ftp/index.html">ensembl</a>
(For <code>fakit subseq</code>)</p>
<ul>
<li><a href="ftp://ftp.ensembl.org/pub/release-84/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz"><code>Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz</code></a></li>
<li><a href="ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.gtf.gz"><code>Homo_sapiens.GRCh38.84.gtf.gz</code></a></li>
<li><code>Homo_sapiens.GRCh38.84.bed.gz</code> is converted from <code>Homo_sapiens.GRCh38.84.gtf.gz</code>
by <a href="http://bedops.readthedocs.org/en/latest/content/reference/file-management/conversion/gtf2bed.html?highlight=gtf2bed"><code>gtf2bed</code></a>
with command<pre><code>zcat Homo_sapiens.GRCh38.84.gtf.gz | gtf2bed --do-not-sort | gzip -c &gt; Homo_sapiens.GRCh38.84.bed.gz
</code></pre>
</li>
</ul>
<p>Only DNA and gtf/bed data of Chr1 were used:</p>
<ul>
<li>
<p><code>chr1.fa.gz</code></p>
<pre><code>    fakit grep -p 1 Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz -o chr1.fa.gz
</code></pre>
</li>
<li>
<p><code>chr1.gtf.gz</code></p>
<pre><code>    zcat Homo_sapiens.GRCh38.84.gtf.gz | grep -w '^1' | gzip -c &gt; chr1.gtf.gz
</code></pre>
</li>
<li>
<p><code>chr1.bed.gz</code></p>
<pre><code>    zcat Homo_sapiens.GRCh38.84.bed.gz | grep -w '^1' | gzip -c &gt; chr1.bed.gz
</code></pre>
</li>
</ul>
<h2 id="seq">seq</h2>
<p>Usage</p>
<pre><code>transform sequences (revserse, complement, extract ID...)

Usage:
  fakit seq [flags]

Flags:
  -p, --complement          complement sequence (blank for Protein sequence)
      --dna2rna             DNA to RNA
  -G, --gap-letter string   gap letters (default &quot;- &quot;)
  -l, --lower-case          print sequences in lower case
  -n, --name                only print names
  -i, --only-id             print ID instead of full head
  -q, --qual                only print qualities
  -g, --remove-gaps         remove gaps
  -r, --reverse             reverse sequence)
      --rna2dna             RNA to DNA
  -s, --seq                 only print sequences
  -u, --upper-case          print sequences in upper case

</code></pre>

<p>Examples</p>
<ol>
<li>
<p>Read and print</p>
<ul>
<li>
<p>From file:</p>
<pre><code>$ fakit seq hairpin.fa.gz
&gt;cel-let-7 MI0000001 Caenorhabditis elegans let-7 stem-loop
UACACUGUGGAUCCGGUGAGGUAGUAGGUUGUAUAGUUUGGAAUAUUACCACCGGUGAAC
UAUGCAAUUUUCUACCUUACCGGAGACAGAACUCUUCGA

$ fakit seq read_1.fq.gz
@HWI-D00523:240:HF3WGBCXX:1:1101:2574:2226 1:N:0:CTGTAG
TGAGGAATATTGGTCAATGGGCGCGAGCCTGAACCAGCCAAGTAGCGTGAAGGATGACTG
CCCTACGGGTTGTAAACTTCTTTTATAAAGGAATAAAGTGAGGCACGTGTGCCTTTTTGT
ATGTACTTTATGAATAAGGATCGGCTAACTCCGTGCCAGCAGCCGCGGTAATACGGAGGA
TCCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGCAGGCGGT
+
HIHIIIIIHIIHGHHIHHIIIIIIIIIIIIIIIHHIIIIIHHIHIIIIIGIHIIIIHHHH
HHGHIHIIIIIIIIIIIGHIIIIIGHIIIIHIIHIHHIIIIHIHHIIIIIIIGIIIIIII
HIIIIIGHIIIIHIIIH?DGHEEGHIIIIIIIIIIIHIIHIIIHHIIHIHHIHCHHIIHG
IHHHHHHH&lt;GG?B@EHDE-BEHHHII5B@GHHF?CGEHHHDHIHIIH
</code></pre>
</li>
<li>
<p>From stdin:</p>
<pre><code>zcat hairpin.fa.gz | fakit seq
</code></pre>
</li>
</ul>
</li>
<li>
<p>Sequence types</p>
<ul>
<li>
<p>In default, <code>fakit seq</code> automatically detect the sequence type</p>
<pre><code>$ echo -e "&gt;seq\nacgtryswkmbdhvACGTRYSWKMBDHV" | fakit stat
file   seq_format   seq_type   num_seqs   min_len   avg_len   max_len
-      FASTA        DNA               1        28        28        28

$ echo -e "&gt;seq\nACGUN ACGUN" | fakit stat
file   seq_format   seq_type   num_seqs   min_len   avg_len   max_len
-      FASTA        RNA               1        11        11        11

$ echo -e "&gt;seq\nabcdefghijklmnpqrstvwyz" | fakit stat
file   seq_format   seq_type   num_seqs   min_len   avg_len   max_len
-      FASTA        Protein           1        23        23        23

$ echo -e "@read\nACTGCN\n+\n@IICCG" | fakit stat
file   seq_format   seq_type   num_seqs   min_len   avg_len   max_len
-      FASTQ        DNA               1         6         6         6
</code></pre>
</li>
<li>
<p>You can also set sequence type by flag <code>-t</code> (<code>--seq-type</code>).
  But this only take effect on subcommands <code>seq</code> and <code>locate</code>.</p>
<pre><code>$ echo -e "&gt;seq\nabcdefghijklmnpqrstvwyz" | fakit seq -t dna
[ERRO] error when parsing seq: seq (invalid DNAredundant letter: e)
</code></pre>
</li>
</ul>
</li>
<li>
<p>Only print names</p>
<ul>
<li>
<p>Full name:</p>
<pre><code>$ fakit seq hairpin.fa.gz -n
cel-let-7 MI0000001 Caenorhabditis elegans let-7 stem-loop
cel-lin-4 MI0000002 Caenorhabditis elegans lin-4 stem-loop
cel-mir-1 MI0000003 Caenorhabditis elegans miR-1 stem-loop
</code></pre>
</li>
<li>
<p>Only ID:</p>
<pre><code>$ fakit seq hairpin.fa.gz -n -i
cel-let-7
cel-lin-4
cel-mir-1
</code></pre>
</li>
<li>
<p>Custom ID region by regular expression (this could be applied to all subcommands):</p>
<pre><code>$ fakit seq hairpin.fa.gz -n -i --id-regexp "^[^\s]+\s([^\s]+)\s"
MI0000001
MI0000002
MI0000003
</code></pre>
</li>
</ul>
</li>
<li>
<p>Only print seq (global flag <code>-w</code> defines the output line width, 0 for no wrap)</p>
<pre><code>$ fakit seq hairpin.fa.gz -s -w 0
UACACUGUGGAUCCGGUGAGGUAGUAGGUUGUAUAGUUUGGAAUAUUACCACCGGUGAACUAUGCAAUUUUCUACCUUACCGGAGACAGAACUCUUCGA
AUGCUUCCGGCCUGUUCCCUGAGACCUCAAGUGUGAGUGUACUAUUGAUGCUUCACACCUGGGCUCUCCGGGUACCAGGACGGUUUGAGCAGAU
AAAGUGACCGUACCGAGCUGCAUACUUCCUUACAUGCCCAUACUAUAUCAUAAAUGGAUAUGGAAUGUAAAGAAGUAUGUAGAACGGGGUGGUAGU
</code></pre>
</li>
<li>
<p>Convert multi-line FASTQ to 4-line FASTQ</p>
<pre><code>$fakit seq read_1.fq.gz -w 0
</code></pre>
</li>
<li>
<p>Reverse comlement sequence</p>
<pre><code>$ fakit seq hairpin.fa.gz -r -p
&gt;cel-let-7 MI0000001 Caenorhabditis elegans let-7 stem-loop
UCGAAGAGUUCUGUCUCCGGUAAGGUAGAAAAUUGCAUAGUUCACCGGUGGUAAUAUUCC
AAACUAUACAACCUACUACCUCACCGGAUCCACAGUGUA
</code></pre>
</li>
<li>
<p>Remove gaps and to lower/upper case</p>
<pre><code>$ echo -e "&gt;seq\nACGT-ACTGC-ACC" | fakit seq -g -u
&gt;seq
ACGTACTGCACC
</code></pre>
</li>
<li>
<p>RNA to DNA</p>
<pre><code>$ echo -e "&gt;seq\nUCAUAUGCUUGUCUCAAAGAUUA" | fakit seq --rna2dna
&gt;seq
TCATATGCTTGTCTCAAAGATTA
</code></pre>
</li>
</ol>
<h2 id="subseq">subseq</h2>
<p>Usage</p>
<pre><code>get subsequences by region/gtf/bed, including flanking sequences.

The definition of region is 1-based and with some custom design.

Examples:

 1-based index    1 2 3 4 5 6 7 8 9 10
negative index    0-9-8-7-6-5-4-3-2-1
           seq    A C G T N a c g t n
           1:1    A
           2:4      C G T
         -4:-2                c g t
         -4:-1                c g t n
         -1:-1                      n
          2:-2      C G T N a c g t
          1:-1    A C G T N a c g t n

Usage:
  fakit subseq [flags]

Flags:
      --bed string        by BED file
  -d, --down-stream int   down stream length
  -T, --feature string    feature type (&quot;.&quot; for all, case ignored) (default &quot;.&quot;)
      --gtf string        by GTF (version 2.2) file
  -f, --only-flank        only return up/down stream sequence
  -r, --region string     by region. e.g 1:12 for first 12 bases, -12:-1 for last 12 bases, 13:-1 for cutting first 12 bases. type &quot;fakit subseq -h&quot; for more examples
  -u, --up-stream int     up stream length

</code></pre>

<p>Examples</p>
<ol>
<li>
<p>First 12 bases</p>
<pre><code>$ zcat hairpin.fa.gz | fakit subseq -r 1:12
</code></pre>
</li>
<li>
<p>Last 12 bases</p>
<pre><code>$ zcat hairpin.fa.gz | fakit subseq -r -12:-1
</code></pre>
</li>
<li>
<p>Subsequences without first and last 12 bases</p>
<pre><code>$ zcat hairpin.fa.gz | fakit subseq -r 13:-13
</code></pre>
</li>
<li>
<p>Get subsequence by GTF file</p>
<pre><code>$ cat t.fa
&gt;seq
actgACTGactgn
$ cat t.gtf
seq     test    CDS     5       8       .       .       .       gene_id "A"; transcript_id "";
seq     test    CDS     5       8       .       -       .       gene_id "B"; transcript_id "";
$ fakit

fakit subseq --gtf t.gtf t.fa
&gt;seq_5:8:. A
ACTG
&gt;seq_5:8:- B
CAGT
</code></pre>
</li>
<li>
<p>Get CDS and 3bp up-stream sequences</p>
<pre><code>$ fakit subseq --gtf t.gtf t.fa -u 3
&gt;seq_5:8:._us:3 A
ctgACTG
&gt;seq_5:8:-_us:3 B
agtCAGT
</code></pre>
</li>
<li>
<p>Get 3bp up-stream sequences of CDS, not including CDS</p>
<pre><code>$ fakit subseq --gtf t.gtf t.fa -u 3 -f
&gt;seq_5:8:._usf:3 A
ctg
&gt;seq_5:8:-_usf:3 B
agt
</code></pre>
</li>
<li>
<p>Get subsequences by BED file. <strong>Note that flag <code>-c 1</code> is used for large genome</strong>.</p>
<pre><code>$ fakit subseq -c 1 --bed chr1.bed.gz chr1.fa.gz -o chr1.bed.gz.fa.gz
</code></pre>
<p>We may need to remove duplicated sequences</p>
<pre><code>$ fakit subseq -c 1 --bed chr1.bed.gz chr1.fa.gz | fakit rmdup -o chr1.bed.gz.rmdup.fa.gz
[INFO] 141060 duplicated records removed
</code></pre>
<p>Summary:</p>
<pre><code>$ fakit stat chr1.bed.gz.*.gz
file                       seq_type    num_seqs    min_len    avg_len      max_len
chr1.bed.gz.fa.gz               DNA     231,974          1    3,089.5    1,551,957
chr1.bed.gz.rmdup.fa.gz         DNA      90,914          1    6,455.8    1,551,957
</code></pre>
</li>
</ol>
<h2 id="sliding">sliding</h2>
<p>Usage</p>
<pre><code>sliding sequences, circle genome supported

Usage:
  fakit sliding [flags]

Flags:
  -C, --circle-genome   circle genome
  -s, --step int        step size
  -W, --window int      window size

</code></pre>

<p>Examples</p>
<ol>
<li>
<p>General use</p>
<pre><code>$ echo -e "&gt;seq\nACGTacgtNN" | fakit sliding -s 3 -W 6
&gt;seq_sliding:1-6
ACGTac
&gt;seq_sliding:4-9
TacgtN
</code></pre>
</li>
<li>
<p>Circle genome</p>
<pre><code>$ echo -e "&gt;seq\nACGTacgtNN" | fakit sliding -s 3 -W 6 -C
&gt;seq_sliding:1-6
ACGTac
&gt;seq_sliding:4-9
TacgtN
&gt;seq_sliding:7-2
gtNNAC
&gt;seq_sliding:10-5
NACGTa
</code></pre>
</li>
<li>
<p>Generate GC content for ploting</p>
<pre><code>$ zcat hairpin.fa.gz | fakit fx2tab | head -n 1 | fakit tab2fx | fakit sliding -s 5 -W 30 | fakit fx2tab -n -g
cel-let-7_sliding:1-30          50.00
cel-let-7_sliding:6-35          46.67
cel-let-7_sliding:11-40         43.33
cel-let-7_sliding:16-45         36.67
cel-let-7_sliding:21-50         33.33
cel-let-7_sliding:26-55         40.00
...
</code></pre>
</li>
</ol>
<h2 id="stat">stat</h2>
<p>Usage</p>
<pre><code>simple statistics of FASTA files

Usage:
  fakit stat [flags]

</code></pre>

<p>Eexamples</p>
<ol>
<li>General use<pre><code>$ fakit stat *.f{a,q}.gz
file            seq_format   seq_type   num_seqs   min_len   avg_len   max_len
clean.fa.gz     FASTA        RNA          26,419        39     104.2     2,354
hairpin.fa.gz   FASTA        RNA          28,645        39       103     2,354
mature.fa.gz    FASTA        RNA          35,828        15      21.8        34
reads_1.fq.gz   FASTQ        DNA           2,500       226       227       229
reads_2.fq.gz   FASTQ        DNA           2,500       223       224       225
</code></pre>
</li>
</ol>
<h2 id="fq2fa">fq2fa</h2>
<p>Usage</p>
<pre><code>covert FASTQ to FASTA

Usage:
  fakit fq2fa [flags]

</code></pre>

<p>Examples</p>
<pre><code>fakit fq2fa reads_1.fq.gz -o reads1_.fa.gz
</code></pre>
<h2 id="fx2tab-tab2fx">fx2tab &amp; tab2fx</h2>
<p>Usage (fx2tab)</p>
<pre><code>covert FASTA/Q to tabular format, and provide various information,
like sequence length, GC content/GC skew.

Usage:
  fakit fx2tab [flags]

Flags:
  -B, --base-content value   print base content. (case ignored, multiple values supported) e.g. -b AT -b N (default [])
  -g, --gc                   print GC content
  -G, --gc-skew              print GC-Skew
  -l, --length               print sequence length
  -n, --name                 only print names (no sequences and qualities)
  -i, --only-id              print ID instead of full head
  -T, --title                print title line

</code></pre>

<p>Usage (tab2fx)</p>
<pre><code>covert tabular format (first two/three columns) to FASTA/Q format

Usage:
  fakit tab2fx [flags]

Flags:
  -p, --comment-line-prefix value   comment line prefix (default [#,//])


</code></pre>

<p>Examples</p>
<ol>
<li>
<p>Default output</p>
<pre><code>$ fakit fx2tab hairpin.fa.gz | head -n 2
cel-let-7 MI0000001 Caenorhabditis elegans let-7 stem-loop      UACACUGUGGAUCCGGUGAGGUAGUAGGUUGUAUAGUUUGGAAUAUUACCACCGGUGAACUAUGCAAUUUUCUACCUUACCGGAGACAGAACUCUUCGA
cel-lin-4 MI0000002 Caenorhabditis elegans lin-4 stem-loop      AUGCUUCCGGCCUGUUCCCUGAGACCUCAAGUGUGAGUGUACUAUUGAUGCUUCACACCUGGGCUCUCCGGGUACCAGGACGGUUUGAGCAGAU
</code></pre>
</li>
<li>
<p>Print sequence length, GC content, and only print names (no sequences),
we could also print title line by flag <code>-T</code>.</p>
<pre><code>$ fakit fx2tab hairpin.fa.gz -l -g -n -i -T | head -n 4 | csvtk -t -C '&amp;' pretty
#name       seq   qual   length   GC
cel-let-7                99       43.43
cel-lin-4                94       54.26
cel-mir-1                96       40.62
</code></pre>
</li>
<li>
<p>Use fx2tab and tab2fx in pipe</p>
<pre><code>$ zcat hairpin.fa.gz | fakit fx2tab | fakit tab2fx

$ zcat reads_1.fq.gz | fakit fx2tab | fakit tab2fx
</code></pre>
</li>
<li>
<p>Sort sequences by length (use <code>fakit sort -l</code>)</p>
<pre><code>$ zcat hairpin.fa.gz | fakit fx2tab -l | sort -t"`echo -e '\t'`" -n -k4,4 | fakit tab2fx
&gt;cin-mir-4129 MI0015684 Ciona intestinalis miR-4129 stem-loop
UUCGUUAUUGGAAGACCUUAGUCCGUUAAUAAAGGCAUC
&gt;mmu-mir-7228 MI0023723 Mus musculus miR-7228 stem-loop
UGGCGACCUGAACAGAUGUCGCAGUGUUCGGUCUCCAGU
&gt;cin-mir-4103 MI0015657 Ciona intestinalis miR-4103 stem-loop
ACCACGGGUCUGUGACGUAGCAGCGCUGCGGGUCCGCUGU

$ fakit sort -l hairpin.fa.gz
</code></pre>
<p>Sorting or filtering by GC (or other base by -flag <code>-B</code>) content could also achieved in similar way.</p>
</li>
<li>
<p>Get first 1000 sequences</p>
<pre><code>$ zcat hairpin.fa.gz | fakit fx2tab | head -n 1000 | fakit tab2fx
</code></pre>
</li>
</ol>
<p><strong>Extension</strong></p>
<p>After converting FASTA to tabular format with <code>fakit fx2tab</code>,
it could be handled with CSV/TSV tools,
 e.g. <a href="https://github.com/shenwei356/csvtkt">csvtk</a>, a cross-platform, efficient and practical CSV/TSV toolkit</p>
<ul>
<li><code>csvtk grep</code> could be used to filter sequences (similar with <code>fakit extract</code>)</li>
<li><code>csvtk inter</code>
computates intersection of multiple files. It could achieve similar function
as <code>fakit common -n</code> along with shell.</li>
<li><code>csvtk join</code> joins multiple CSV/TSV files by multiple IDs.</li>
<li><a href="https://github.com/shenwei356/datakit/blob/master/csv_melt">csv_melt</a>
provides melt function, could be used in preparation of data for ploting.</li>
</ul>
<h2 id="grep">grep</h2>
<p>Usage</p>
<pre><code>search sequences by pattern(s) of name or sequence motifs

Usage:
  fakit grep [flags]

Flags:
  -n, --by-name               match by full name instead of just id
  -s, --by-seq                match by seq
  -d, --degenerate            pattern/motif contains degenerate base
      --delete-matched        delete matched pattern to speedup
  -i, --ignore-case           ignore case
  -v, --invert-match          invert the sense of matching, to select non-matching records
  -p, --pattern value         search pattern (multiple values supported) (default [])
  -f, --pattern-file string   pattern file
  -r, --use-regexp            patterns are regular expression

</code></pre>

<p>Examples</p>
<ol>
<li>
<p>Extract human hairpins (i.e. sequences with name starting with <code>hsa</code>)</p>
<pre><code>$ zcat hairpin.fa.gz | fakit grep -r -p ^hsa
&gt;hsa-let-7a-1 MI0000060 Homo sapiens let-7a-1 stem-loop
UGGGAUGAGGUAGUAGGUUGUAUAGUUUUAGGGUCACACCCACCACUGGGAGAUAACUAU
ACAAUCUACUGUCUUUCCUA
&gt;hsa-let-7a-2 MI0000061 Homo sapiens let-7a-2 stem-loop
AGGUUGAGGUAGUAGGUUGUAUAGUUUAGAAUUACAUCAAGGGAGAUAACUGUACAGCCU
CCUAGCUUUCCU
</code></pre>
</li>
<li>
<p>Remove human and mice hairpins.</p>
<pre><code>$ zcat hairpin.fa.gz | fakit grep -r -p ^hsa -p ^mmu -v
</code></pre>
</li>
<li>
<p>Extract new entries by information from miRNA.diff.gz</p>
<ol>
<li>
<p>Get IDs of new entries.</p>
<pre><code>$ zcat miRNA.diff.gz | grep ^# -v | grep NEW | cut -f 2 &gt; list
$ more list
cfa-mir-486
cfa-mir-339-1
pmi-let-7
</code></pre>
</li>
<li>
<p>Extract by ID list file</p>
<pre><code>$ zcat hairpin.fa.gz | fakit grep -f list &gt; new.fa
</code></pre>
</li>
</ol>
</li>
<li>
<p>Extract sequences starting with AGGCG</p>
<pre><code>$ zcat hairpin.fa.gz | fakit grep -s -r -i -p ^aggcg
</code></pre>
</li>
<li>
<p>Extract sequences with TTSAA (AgsI digest site) in SEQUENCE. Base S stands for C or G.</p>
<pre><code>$ zcat hairpin.fa.gz | fakit grep -s -d -i -p TTSAA
</code></pre>
<p>It's equal to but simpler than:</p>
<pre><code>$ zcat hairpin.fa.gz | fakit grep -s -r -i -p TT[CG]AA
</code></pre>
</li>
</ol>
<h2 id="locate">locate</h2>
<p>Usage</p>
<pre><code>locate subsequences/motifs

Motifs could be EITHER plain sequence containing &quot;ACTGN&quot; OR regular
expression like &quot;A[TU]G(?:.{3})+?[TU](?:AG|AA|GA)&quot; for ORFs.
Degenerate bases like &quot;RYMM..&quot; are also supported by flag -d.

In default, motifs are treated as regular expression.
When flag -d given, regular expression may be wrong.
For example: &quot;\w&quot; will be wrongly converted to &quot;\[AT]&quot;.

Usage:
  fakit locate [flags]

Flags:
  -d, --degenerate             pattern/motif contains degenerate base
  -i, --ignore-case            ignore case
  -P, --only-positive-strand   only search at positive strand
  -p, --pattern value          search pattern/motif (multiple values supported) (default [])
  -f, --pattern-file string    pattern/motif file (FASTA format)

</code></pre>

<p>Examples</p>
<ol>
<li>
<p>Locate ORFs.</p>
<pre><code>$ zcat hairpin.fa.gz | fakit locate -i -p "A[TU]G(?:.{3})+?[TU](?:AG|AA|GA)"
seqID   patternName     pattern strand  start   end     matched
cel-lin-4       A[TU]G(?:.{3})+?[TU](?:AG|AA|GA)        A[TU]G(?:.{3})+?[TU](?:AG|AA|GA)        +  136      AUGCUUCCGGCCUGUUCCCUGAGACCUCAAGUGUGA
cel-mir-1       A[TU]G(?:.{3})+?[TU](?:AG|AA|GA)        A[TU]G(?:.{3})+?[TU](?:AG|AA|GA)        +  54       95      AUGGAUAUGGAAUGUAAAGAAGUAUGUAGAACGGGGUGGUAG
cel-mir-1       A[TU]G(?:.{3})+?[TU](?:AG|AA|GA)        A[TU]G(?:.{3})+?[TU](?:AG|AA|GA)        -  43       51      AUGAUAUAG
</code></pre>
</li>
<li>
<p>Locate Motif.</p>
<pre><code>$ zcat hairpin.fa.gz | fakit locate -i -p UUS
seqID   patternName     pattern strand  start   end     matched
bna-MIR396a     UUS     UUS     -       105     107     UUS
bna-MIR396a     UUS     UUS     -       89      91      UUS
</code></pre>
<p>Notice that <code>fakit grep</code> only searches in positive strand, but <code>fakit loate</code> could recognize both strand</p>
</li>
</ol>
<h2 id="rmdup">rmdup</h2>
<p>Usage</p>
<pre><code>remove duplicated sequences by id/name/sequence

Usage:
  fakit rmdup [flags]

Flags:
    -n, --by-name                by full name instead of just id
    -s, --by-seq                 by seq
    -D, --dup-num-file string    file to save number and list of duplicated seqs
    -d, --dup-seqs-file string   file to save duplicated seqs
    -i, --ignore-case            ignore case
    -m, --md5                    use MD5 instead of original seqs to reduce memory usage when comparing by seqs

</code></pre>

<p>Examples</p>
<p>Similar to <code>common</code>.</p>
<ol>
<li>
<p>General use</p>
<pre><code>$ zcat hairpin.fa.gz | fakit rmdup -s -o clean.fa.gz
[INFO] 2226 duplicated records removed

$ zcat reads_1.fq.gz | fakit rmdup -s -o clean.fa.gz
[INFO] 1086 duplicated records removed
</code></pre>
</li>
<li>
<p>Save duplicated sequences to file</p>
<pre><code>$ zcat hairpin.fa.gz | fakit rmdup -s -i -m -o clean.fa.gz -d duplicated.fa.gz -D duplicated.detail.txt

$ cat duplicated.detail.txt   # here is not the entire list
3   hsa-mir-424, mml-mir-424, ppy-mir-424
3   hsa-mir-342, mml-mir-342, ppy-mir-342
2   ngi-mir-932, nlo-mir-932
2   ssc-mir-9784-1, ssc-mir-9784-2
</code></pre>
</li>
</ol>
<h2 id="common">common</h2>
<p>Usage</p>
<pre><code>find common sequences of multiple files by id/name/sequence

Usage:
  fakit common [flags]

Flags:
    -n, --by-name       match by full name instead of just id
    -s, --by-seq        match by sequence
    -i, --ignore-case   ignore case
    -m, --md5           use MD5 instead of original seqs to reduce memory usage when comparing by seqs

</code></pre>

<p>Examples</p>
<ol>
<li>
<p>By ID (default)</p>
<pre><code>fakit common file*.fa -o common.fasta
</code></pre>
</li>
<li>
<p>By full name</p>
<pre><code>fakit common file*.fa -n -o common.fasta
</code></pre>
</li>
<li>
<p>By sequence</p>
<pre><code>fakit common file*.fa -s -i -o common.fasta
</code></pre>
</li>
<li>
<p>By sequence (large sequences)</p>
<pre><code>fakit common file*.fa -s -i -o common.fasta -m
</code></pre>
</li>
</ol>
<h2 id="split">split</h2>
<p>Usage</p>
<pre><code>split sequences into files by name ID, subsequence of given region,
part size or number of parts.

The definition of region is 1-based and with some custom design.

Examples:

 0-based index    0 1 2 3 4 5 6 7 8 9
 1-based index    1 2 3 4 5 6 7 8 9 10
negative index    0-9-8-7-6-5-4-3-2-1
           seq    A C G T N a c g t n
           1:1    A
           2:4        G T N
         -4:-2                c g t
         -4:-1                c g t n
         -1:-1                      n
          2:-2      C G T N a c g t
          1:-1    A C G T N a c g t n

Usage:
  fakit split [flags]

Flags:
  -i, --by-id              split squences according to sequence ID
  -p, --by-part int        split squences into N parts
  -r, --by-region string   split squences according to subsequence of given region. e.g 1:12 for first 12 bases, -12:-1 for last 12 bases. type &quot;fakit split -h&quot; for more example
  -s, --by-size int        split squences into multi parts with N sequences
  -d, --dry-run            dry run, just print message and no files will be created.
  -m, --md5                use MD5 instead of region sequence in output file when using flag -r (--by-region)
  -2, --two-pass           2-pass mode read files twice to lower memory usage. Not allowed when reading from stdin

</code></pre>

<p>Examples</p>
<ol>
<li>
<p>Split sequences into parts with at most 10000 sequences</p>
<pre><code>$ fakit split hairpin.fa.gz -s 10000
[INFO] split into 10000 seqs per file
[INFO] write 10000 sequences to file: hairpin.fa.part_001.gz
[INFO] write 10000 sequences to file: hairpin.fa.part_002.gz
[INFO] write 8645 sequences to file: hairpin.fa.part_003.gz
</code></pre>
</li>
<li>
<p>Split sequences into 4 parts</p>
<pre><code>$ fakit split hairpin.fa.gz -p 4
[INFO] split into 4 parts
[INFO] read sequences ...
[INFO] read 28645 sequences
[INFO] write 7162 sequences to file: hairpin.fa.part_001.gz
[INFO] write 7162 sequences to file: hairpin.fa.part_002.gz
[INFO] write 7162 sequences to file: hairpin.fa.part_003.gz
[INFO] write 7159 sequences to file: hairpin.fa.part_004.gz
</code></pre>
<p>To reduce memory usage when spliting big file, we could use flag <code>--two-pass</code></p>
<pre><code>$ fakit split hairpin.fa.gz -p 4 -2
[INFO] split into 4 parts
[INFO] first pass: get seq number
[INFO] seq number: 28645
[INFO] second pass: read and split
[INFO] write 7162 sequences to file: hairpin.fa.part_001.gz
[INFO] write 7162 sequences to file: hairpin.fa.part_002.gz
[INFO] write 7162 sequences to file: hairpin.fa.part_003.gz
[INFO] write 7159 sequences to file: hairpin.fa.part_004.gz
</code></pre>
</li>
<li>
<p>Split sequences by species. i.e. by custom IDs (first three letters)</p>
<pre><code>$ fakit split hairpin.fa.gz -i --id-regexp "^([\w]+)\-"
[INFO] split by ID. idRegexp: ^([\w]+)\-
[INFO] read sequences ...
[INFO] read 28645 sequences
[INFO] write 97 sequences to file: hairpin.fa.id_asu.gz
[INFO] write 267 sequences to file: hairpin.fa.id_chi.gz
[INFO] write 296 sequences to file: hairpin.fa.id_gra.gz
...
</code></pre>
</li>
<li>
<p>Split sequences by sequence region (for example, sequence barcode)</p>
<pre><code>$ fakit split hairpin.fa.gz -r 1:12
[INFO] split by region: ^([^\s]+)\s?
[INFO] read sequences ...
[INFO] read 28645 sequences
[INFO] write 1 sequences to file: hairpin.fa.region_1:12_UGUUUGCUCAGC.gz
[INFO] write 1 sequences to file: hairpin.fa.region_1:12_GAAGAAGAAGAC.gz
[INFO] write 4 sequences to file: hairpin.fa.region_1:12_UGAGUGUAGUGC.gz
</code></pre>
<p>If region is too long, we could use falg <code>-m</code>, i.e. use MD5 instead of region sequence in output file.</p>
<p>Sequence suffix could be defined as <code>-r -12:-1</code></p>
</li>
</ol>
<h2 id="sample">sample</h2>
<p>Usage</p>
<pre><code>sample sequences by number or proportion.

Usage:
  fakit sample [flags]

Flags:
  -n, --number int         sample by number (result may not exactly match)
  -p, --proportion float   sample by proportion
  -s, --rand-seed int      rand seed for shuffle (default 11)
  -2, --two-pass           2-pass mode read files twice to lower memory usage. Not allowed when reading from stdin

</code></pre>

<p>Examples</p>
<ol>
<li>
<p>Sample by number</p>
<pre><code>$ zcat hairpin.fa.gz | fakit sample -n 1000 -o sample.fa.gz
[INFO] sample by number
[INFO] 949 sequences outputed
</code></pre>
<p>To reduce memory usage when spliting big file, we could use flag <code>--two-pass</code></p>
</li>
<li>
<p>Sample by proportion</p>
<pre><code>$ zcat hairpin.fa.gz | fakit sample -p 0.1 -o sample.fa.gz
[INFO] sample by proportion
[INFO] 2814 sequences outputed
</code></pre>
</li>
<li>
<p>Set rand seed to reproduce the result</p>
<pre><code>$ zcat hairpin.fa.gz | fakit sample -p 0.1 -s 11
</code></pre>
</li>
<li>
<p>Most of the time, we could shuffle after sampling</p>
<pre><code>$ zcat hairpin.fa.gz | fakit sample -p 0.1 | fakit shuffle -o sample.fa.gz
</code></pre>
</li>
</ol>
<h2 id="replace">replace</h2>
<p>Usage</p>
<pre><code>replace name/sequence/by regular expression.

Note that the replacement supports capture variables.
e.g. $1 represents the text of the first submatch.
ATTENTION: use SINGLE quote NOT double quotes in *nix OS.

Examples: Adding space to all bases.

    fakit replace -p &quot;(.)&quot; -r '$1 ' -s

Or use the \ escape character.

    fakit replace -p &quot;(.)&quot; -r &quot;\$1 &quot; -s

more on: http://shenwei356.github.io/fakit/usage/#replace

Usage:
  fakit replace [flags]

Flags:
  -s, --by-seq               replace seq
  -i, --ignore-case          ignore case
  -p, --pattern string       search regular expression
  -r, --replacement string   replacement. supporting capture variables.  e.g. $1 represents the text of the first submatch. ATTENTION: use SINGLE quote NOT double quotes in *nix OS or use the \ escape character.

</code></pre>

<p>Examples</p>
<ol>
<li>
<p>Remove descriptions</p>
<pre><code>$ echo -e "&gt;seq1 abc-123\nACGT-ACGT" | fakit replace -p " .+"
&gt;seq1
ACGT-ACGT
</code></pre>
</li>
<li>
<p>Replace "-" with "="</p>
<pre><code>$ echo -e "&gt;seq1 abc-123\nACGT-ACGT" | fakit replace -p "\-" -r '='
&gt;seq1 abc=123
ACGT-ACGT
</code></pre>
</li>
<li>
<p>Remove gaps in sequences.</p>
<pre><code>$ echo -e "&gt;seq1 abc-123\nACGT-ACGT" | fakit replace -p " |-" -s
&gt;seq1 abc-123
ACGTACGT
</code></pre>
</li>
<li>
<p>Add space to every base. <strong>ATTENTION: use SINGLE quote NOT double quotes in *nix OS</strong></p>
<pre><code>$ echo -e "&gt;seq1 abc-123\nACGT-ACGT" | fakit replace -p "(.)" -r '$1 ' -s
&gt;seq1 abc-123
A C G T - A C G T
</code></pre>
</li>
<li>
<p>Transpose sequence with <a href="https://github.com/shenwei356/csvtk">csvtk</a></p>
<pre><code>$ echo -e "&gt;seq1\nACTGACGT\n&gt;seq2\nactgccgt" | fakit replace -p "(.)" -r     "\$1 " -s | fakit seq -s -u | csvtk space2tab | csvtk -t transpose
A       A
C       C
T       T
G       G
A       C
C       C
G       G
T       T
</code></pre>
</li>
</ol>
<h2 id="shuffle">shuffle</h2>
<p>Usage</p>
<pre><code>shuffle sequences

Usage:
  fakit shuffle [flags]

Flags:
  -s, --rand-seed int   rand seed for shuffle (default 23)

</code></pre>

<p>Examples</p>
<ol>
<li>General use.<pre><code>$ zcat hairpin.fa.gz | fakit shuffle -o shuffled.fa.gz
[INFO] read sequences ...
[INFO] 28645 sequences loaded
[INFO] shuffle ...
[INFO] output ...
</code></pre>
</li>
</ol>
<h2 id="sort">sort</h2>
<p>Usage</p>
<pre><code>sort sequences by id/name/sequence/length

Usage:
  fakit sort [flags]

Flags:
  -l, --by-length     by sequence length
  -n, --by-name       by full name instead of just id
  -s, --by-seq        by sequence
  -i, --ignore-case   ignore case
  -r, --reverse       reverse the result

</code></pre>

<p>Examples</p>
<ol>
<li>
<p>sort by ID</p>
<pre><code>$ echo -e "&gt;seq1\nACGTNcccc\n&gt;SEQ2\nacgtnAAAA" | fakit sort --quiet
&gt;SEQ2
acgtnAAAA
&gt;seq1
ACGTNcccc
</code></pre>
</li>
<li>
<p>sort by ID, ignoring case.</p>
<pre><code>$ echo -e "&gt;seq1\nACGTNcccc\n&gt;SEQ2\nacgtnAAAA" | fakit sort --quiet -i
&gt;seq1
ACGTNcccc
&gt;SEQ2
acgtnAAAA
</code></pre>
</li>
<li>
<p>sort by seq, ignoring case.</p>
<pre><code>$ echo -e "&gt;seq1\nACGTNcccc\n&gt;SEQ2\nacgtnAAAA" | fakit sort --quiet -s -i
&gt;SEQ2
acgtnAAAA
&gt;seq1
ACGTNcccc
</code></pre>
</li>
<li>
<p>sort by sequence length</p>
<pre><code>$ echo -e "&gt;seq1\nACGTNcccc\n&gt;SEQ2\nacgtnAAAAnnn\n&gt;seq3\nacgt" | fakit sort --quiet -l
&gt;seq3
acgt
&gt;seq1
ACGTNcccc
&gt;SEQ2
acgtnAAAAnnn
</code></pre>
</li>
</ol>
<div id="disqus_thread"></div>

<script>
/**
* RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
* LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables
*/
/*
var disqus_config = function () {
this.page.url = PAGE_URL; // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');

s.src = '//fastakit.disqus.com/embed.js';

s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>

<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript" rel="nofollow">comments powered by Disqus.</a></noscript></div>
            
        </div>

        <footer class="col-md-12">
            <hr>
            
            <p>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>

        <script src="../js/jquery-1.10.2.min.js"></script>
        <script src="../js/bootstrap-3.0.3.min.js"></script>
        <script src="../js/highlight.pack.js"></script>
        <script>var base_url = '..';</script>
        <script data-main="../mkdocs/js/search.js" src="../mkdocs/js/require.js"></script>
        <script src="../js/base.js"></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
                        <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                    </div>
                    <div class="modal-body">
                        <p>
                            From here you can search these documents. Enter
                            your search terms below.
                        </p>
                        <form role="form">
                            <div class="form-group">
                                <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                            </div>
                        </form>
                        <div id="mkdocs-search-results"></div>
                    </div>
                    <div class="modal-footer">
                    </div>
                </div>
            </div>
        </div>

    </body>
</html>
