<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="SeqKit - a cross-platform and ultrafast toolkit for FASTA/Q file manipulation"> 
    <meta name="author" content="Wei Shen"> 
    <link rel="shortcut icon" href="../img/favicon.ico">

    <title>Download - SeqKit - Ultrafast FASTA/Q kit</title>

    <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/font-hack/2.018/css/hack.min.css">
    <link href='//fonts.googleapis.com/css?family=PT+Sans:400,400italic,700,700italic&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../css/base.css" rel="stylesheet">
    <link href="../css/cinder.css" rel="stylesheet">
    <link rel="stylesheet" href="../css/highlight.css">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

    <script src="https://ajax.googleapis.com/ajax/libs/webfont/1.5.18/webfont.js"></script>
    <script>
    WebFont.load({
        google: {
            families: ['Open Sans', 'PT Sans']
        }
    });
    </script>

    
    <script>
    (function(i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r;
        i[r] = i[r] || function() {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date();
        a = s.createElement(o),
        m = s.getElementsByTagName(o)[0];
        a.async = 1;
        a.src = g;
        m.parentNode.insertBefore(a, m)
    })(window, document, 'script', '//www.google-analytics.com/analytics.js', 'ga');

    ga('create', 'UA-28948626-9', 'bioinf.shenwei.me/seqkit');
    ga('send', 'pageview');
    </script>
    
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->

            <a class="navbar-brand" href="..">SeqKit - Ultrafast FASTA/Q kit</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li class="active">
                        <a href="./">Download</a>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Documents <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../usage/">Usage</a>
</li>

                        
                            
<li >
    <a href="../faq/">FAQ</a>
</li>

                        
                            
<li >
    <a href="../tutorial/">Tutorial</a>
</li>

                        
                            
<li >
    <a href="../benchmark/">Benchmark</a>
</li>

                        
                            
<li >
    <a href="../note/">Notes</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Links <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../bioinf/">More tools</a>
</li>

                        
                            
<li >
    <a href="../yanlilab/">YanLi Lab</a>
</li>

                        
                        </ul>
                    </li>
                
                
                </ul>

            <ul class="nav navbar-nav navbar-right">
                    <li>
                        <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                            <i class="fa fa-search"></i> Search
                        </a>
                    </li>
                    <li >
                        <a rel="next" href="..">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../usage/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                    <li>
                        <a href="https://github.com/shenwei356/seqkit/edit/master/docs/download.md"><i class="fa fa-github"></i> Edit on GitHub</a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#download">Download</a></li>
            <li class="second-level"><a href="#latest-version">Latest Version</a></li>
                
                <li class="third-level"><a href="#please-cite">Please cite</a></li>
                <li class="third-level"><a href="#links">Links</a></li>
            <li class="second-level"><a href="#installation">Installation</a></li>
                
                <li class="third-level"><a href="#method-1-download-binaries-latest-stabledev-version">Method 1: Download binaries (latest stable/dev version)</a></li>
                <li class="third-level"><a href="#method-2-install-via-cond-latest-stable-version">Method 2: Install via cond (latest stable version)</a></li>
                <li class="third-level"><a href="#method-3-install-via-homebrew-latest-stable-version">Method 3: Install via homebrew (latest stable version)</a></li>
                <li class="third-level"><a href="#method-4-for-go-developer-latest-stabledev-version">Method 4: For Go developer (latest stable/dev version)</a></li>
                <li class="third-level"><a href="#method-5-docker-based-installation-latest-stabledev-version">Method 5: Docker based installation (latest stable/dev version)</a></li>
            <li class="second-level"><a href="#bash-completion">Bash-completion</a></li>
                
            <li class="second-level"><a href="#release-history">Release History</a></li>
                
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<h1 id="download">Download</h1>
<p>SeqKit is implemented in <a href="https://golang.org/">Go</a> programming language,
 executable binary files <strong>for most popular operating systems</strong> are freely available
  in <a href="https://github.com/shenwei356/seqkit/releases">release</a> page.</p>
<h2 id="latest-version">Latest Version</h2>
<ul>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.9.2">SeqKit v0.9.2</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.9.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.9.2/total.svg" /></a><ul>
<li><code>seqkit</code>: stricter checking for value of global flag <code>-t/--seq-type</code>.</li>
<li><code>seqkit sliding</code>: fix bug for flag <code>-g/--greedy</code>. <a href="https://github.com/shenwei356/seqkit/issues/54">#54</a></li>
<li><code>seqkit translate</code>: fix bug for frame &lt; 0. <a href="https://github.com/shenwei356/seqkit/issues/55">#55</a></li>
<li><code>seqkit seq</code>: add TAB to default blank characters (flag <code>-G/--gap-letters</code>), and fix filter result when using flag <code>-g/--remove-gaps</code> along with <code>-m/--min-len</code> or <code>-M/--max-len</code></li>
</ul>
</li>
</ul>
<h3 id="please-cite">Please cite</h3>
<p><strong>W Shen</strong>, S Le, Y Li*, F Hu*. SeqKit: a cross-platform and ultrafast toolkit for FASTA/Q file manipulation.
<strong><em>PLOS ONE</em></strong>. <a href="https://doi.org/10.1371/journal.pone.0163962">doi:10.1371/journal.pone.0163962</a>.</p>
<h3 id="links">Links</h3>
<p><strong><em>Tips</em></strong></p>
<ul>
<li>run <code>seqkit version</code> to check update !!!</li>
<li>run <code>seqkit genautocomplete</code> to update Bash completion !!!</li>
</ul>
<table>
<thead>
<tr>
<th align="left">OS</th>
<th align="left">Arch</th>
<th align="left">File, 中国镜像</th>
<th align="left">Download Count</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Linux</td>
<td align="left">32-bit</td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_linux_386.tar.gz">seqkit_linux_386.tar.gz</a>, <br/> <a href="http://app.shenwei.me/data/seqkit/seqkit_linux_386.tar.gz">中国镜像</a></td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_linux_386.tar.gz"><img alt="Github Releases (by Asset)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/latest/seqkit_linux_386.tar.gz.svg?maxAge=3600" /></a></td>
</tr>
<tr>
<td align="left">Linux</td>
<td align="left"><strong>64-bit</strong></td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_linux_amd64.tar.gz"><strong>seqkit_linux_amd64.tar.gz</strong></a>, <br/> <a href="http://app.shenwei.me/data/seqkit/seqkit_linux_amd64.tar.gz">中国镜像</a></td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_linux_amd64.tar.gz"><img alt="Github Releases (by Asset)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/latest/seqkit_linux_amd64.tar.gz.svg?maxAge=3600" /></a></td>
</tr>
<tr>
<td align="left">OS X</td>
<td align="left">32-bit</td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_darwin_386.tar.gz">seqkit_darwin_386.tar.gz</a>, <br/> <a href="http://app.shenwei.me/data/seqkit/seqkit_darwin_386.tar.gz">中国镜像</a></td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_darwin_386.tar.gz"><img alt="Github Releases (by Asset)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/latest/seqkit_darwin_386.tar.gz.svg?maxAge=3600" /></a></td>
</tr>
<tr>
<td align="left">OS X</td>
<td align="left"><strong>64-bit</strong></td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_darwin_amd64.tar.gz"><strong>seqkit_darwin_amd64.tar.gz</strong></a>, <br/> <a href="http://app.shenwei.me/data/seqkit/seqkit_darwin_amd64.tar.gz">中国镜像</a></td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_darwin_amd64.tar.gz"><img alt="Github Releases (by Asset)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/latest/seqkit_darwin_amd64.tar.gz.svg?maxAge=3600" /></a></td>
</tr>
<tr>
<td align="left">Windows</td>
<td align="left">32-bit</td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_windows_386.exe.tar.gz">seqkit_windows_386.exe.tar.gz</a>, <br/> <a href="http://app.shenwei.me/data/seqkit/seqkit_windows_386.exe.tar.gz">中国镜像</a></td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_windows_386.exe.tar.gz"><img alt="Github Releases (by Asset)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/latest/seqkit_windows_386.exe.tar.gz.svg?maxAge=3600" /></a></td>
</tr>
<tr>
<td align="left">Windows</td>
<td align="left"><strong>64-bit</strong></td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_windows_amd64.exe.tar.gz"><strong>seqkit_windows_amd64.exe.tar.gz</strong></a>, <br/> <a href="http://app.shenwei.me/data/seqkit/seqkit_windows_amd64.exe.tar.gz">中国镜像</a></td>
<td align="left"><a href="https://github.com/shenwei356/seqkit/releases/download/v0.9.2/seqkit_windows_amd64.exe.tar.gz"><img alt="Github Releases (by Asset)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/latest/seqkit_windows_amd64.exe.tar.gz.svg?maxAge=3600" /></a></td>
</tr>
</tbody>
</table>
<h2 id="installation">Installation</h2>
<h4 id="method-1-download-binaries-latest-stabledev-version">Method 1: Download binaries (latest stable/dev version)</h4>
<p>Just <a href="https://github.com/shenwei356/seqkit/releases">download</a> compressed
executable file of your operating system,
and decompress it with <code>tar -zxvf *.tar.gz</code> command or other tools.
And then:</p>
<ol>
<li>
<p><strong>For Linux-like systems</strong></p>
<ol>
<li>
<p>If you have root privilege simply copy it to <code>/usr/local/bin</code>:</p>
<pre><code>sudo cp seqkit /usr/local/bin/
</code></pre>
</li>
<li>
<p>Or copy to anywhere in the environment variable <code>PATH</code>:</p>
<pre><code>mkdir -p $HOME/bin/; cp seqkit $HOME/bin/
</code></pre>
</li>
</ol>
</li>
<li>
<p><strong>For windows</strong>, just copy <code>seqkit.exe</code> to <code>C:\WINDOWS\system32</code>.</p>
</li>
</ol>
<h4 id="method-2-install-via-cond-latest-stable-version">Method 2: Install via cond (latest stable version) <a href="http://bioinf.shenwei.me/seqkit/download/"><img alt="Install-with-conda" src="https://anaconda.org/bioconda/seqkit/badges/installer/conda.svg" /></a> <a href="https://anaconda.org/bioconda/seqkit"><img alt="Anaconda Cloud" src="https://anaconda.org/bioconda/seqkit/badges/version.svg" /></a> <a href="https://anaconda.org/bioconda/seqkit"><img alt="downloads" src="https://anaconda.org/bioconda/seqkit/badges/downloads.svg" /></a></h4>
<pre><code>conda install -c bioconda seqkit
</code></pre>
<h4 id="method-3-install-via-homebrew-latest-stable-version">Method 3: Install via homebrew (latest stable version)</h4>
<pre><code>brew install brewsci/bio/seqkit
</code></pre>
<h4 id="method-4-for-go-developer-latest-stabledev-version">Method 4: For Go developer (latest stable/dev version)</h4>
<pre><code>go get -u github.com/shenwei356/seqkit/seqkit
</code></pre>
<h4 id="method-5-docker-based-installation-latest-stabledev-version">Method 5: Docker based installation (latest stable/dev version)</h4>
<p><a href="https://docs.docker.com/engine/installation/#supported-platforms">Install Docker</a></p>
<p>git clone this repo:</p>
<pre><code>git clone https://github.com/shenwei356/seqkit
</code></pre>
<p>Run the following commands:</p>
<pre><code>cd seqkit
docker build -t shenwei356/seqkit .
docker run -it shenwei356/seqkit:latest
</code></pre>
<h2 id="bash-completion">Bash-completion</h2>
<p>Note: The current version supports Bash only.
This should work for *nix systems with Bash installed.</p>
<p>Howto:</p>
<ol>
<li>
<p>run: <code>seqkit genautocomplete</code></p>
</li>
<li>
<p>create and edit <code>~/.bash_completion</code> file if you don't have it.</p>
<pre><code>nano ~/.bash_completion
</code></pre>
<p>add the following:</p>
<pre><code>for bcfile in ~/.bash_completion.d/* ; do
  . $bcfile
done
</code></pre>
</li>
</ol>
<h2 id="release-history">Release History</h2>
<ul>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.9.1">SeqKit v0.9.1</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.9.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.9.1/total.svg" /></a><ul>
<li><code>seqkit faidx</code>: fix bug of retrieving subsequence with multiple regions on same sequence. <a href="https://github.com/shenwei356/seqkit/issues/48">#48</a></li>
<li><code>seqkit sort/shuffle/split</code>: fix bug when using <code>-2/--two-pass</code> to process <code>.gz</code> files. <a href="https://github.com/shenwei356/seqkit/issues/52">#52</a></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.9.0">SeqKit v0.9.0</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.9.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.9.0/total.svg" /></a><ul>
<li><code>seqkit</code>: better handle of empty file, no error message shown. <a href="https://github.com/shenwei356/seqkit/issues/36">#36</a></li>
<li>new subcommand <code>seqkit split2</code>: split sequences into files by size/parts (FASTA, PE/SE FASTQ).  <a href="https://github.com/shenwei356/seqkit/issues/35">#35</a></li>
<li>new subcommand <code>seqkit translate</code>: translate DNA/RNA to protein sequence. <a href="https://github.com/shenwei356/seqkit/issues/28">#28</a></li>
<li><code>seqkit sort</code>: fix bug when using <code>-2 -i</code>, and add support for sorting in natural order. <a href="https://github.com/shenwei356/seqkit/issues/39">#39</a></li>
<li><code>seqkit grep</code> and <code>seqkit locate</code>: add experimental support of mismatch when searching subsequences. <a href="https://github.com/shenwei356/seqkit/issues/14">#14</a></li>
<li><code>seqkit stats</code>: add stats of Q20 and Q30 for FASTQ. <a href="https://github.com/shenwei356/seqkit/issues/45">#45</a></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.8.1">SeqKit v0.8.1</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.8.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.8.1/total.svg" /></a><ul>
<li><code>seqkit</code>: do not call <code>pigz</code> or <code>gzip</code> for decompressing gzipped file any more. But you can still utilize <code>pigz</code> or <code>gzip</code> by <code>pigz -d -c seqs.fq.gz | seqkit xxx</code>.</li>
<li><code>seqkit subseq</code>: fix bug of missing quality when using <code>--gtf</code> or <code>--bed</code></li>
<li><code>seqkit stats</code>: parallelize counting files, it's much faster for lots of small files, especially for files on SSD</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.8.0">SeqKit v0.8.0</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.8.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.8.0/total.svg" /></a><ul>
<li><code>seqkit</code>, <strong>stricter FASTA/Q format requirement</strong>, i.e., must starting with <code>&gt;</code> or <code>@</code>.</li>
<li><code>seqkit</code>, <em>fix output format for FASTQ files containing zero-length records</em>, yes this <a href="https://github.com/lh3/seqtk/issues/109">happens</a>.</li>
<li><code>seqkit</code>, add amino acid code <code>O</code> (pyrrolysine) and <code>U</code> (selenocysteine).</li>
<li><code>seqkit replace</code>, <em>add flag <code>--nr-width</code> to fill leading 0s for <code>{nr}</code></em>, useful for preparing sequence submission ("&gt;strain_00001 XX", "&gt;strain_00002 XX").</li>
<li><code>seqkit subseq</code>, require BED file to be tab-delimited.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.7.2">SeqKit v0.7.2</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.7.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.7.2/total.svg" /></a><ul>
<li><code>seqkit tab2fx</code>: fix a concurrency bug that occurs in low proprobability
when only 1-column data provided.</li>
<li><code>seqkit stats</code>: add quartiles of sequence length</li>
<li><code>seqkit faidx</code>: add support for retrieving subsequence using seq ID and region,
which is similar with "samtools faidx" but has some extra features</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.7.1">SeqKit v0.7.1</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.7.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.7.1/total.svg" /></a><ul>
<li><code>seqkit convert</code>: fix bug of read quality containing only 3 or less values.  <a href="https://github.com/shenwei356/bio/issues/3">shenwei356/bio/issues/3</a></li>
<li><code>seqkit stats</code>: add option <code>-T/--tabular</code> to output in machine-friendly tabular format.   <a href="https://github.com/shenwei356/seqkit/issues/23">#23</a></li>
<li><code>seqkit common</code>: increase speed and decrease memory occupation, and add some notes.</li>
<li>fix some typos. <a href="https://github.com/shenwei356/seqkit/issues/22">#22</a></li>
<li>suggestion: please <strong>install <a href="http://zlib.net/pigz/">pigz</a> to gain better parsing performance for gzipped data</strong>.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.7.0">SeqKit v0.7.0</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.7.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.7.0/total.svg" /></a><ul>
<li>add new command <code>convert</code> for converting FASTQ quality encoding between Sanger, Solexa and Illumina. Thanks suggestion from <a href="https://github.com/cviner">@cviner</a> ( <a href="https://github.com/shenwei356/seqkit/issues/18">#18</a>). <a href="http://bioinf.shenwei.me/seqkit/usage/#convert">usage &amp; example</a>.</li>
<li>add new command <code>range</code> for printing FASTA/Q records in a range (start:end). <a href="https://github.com/shenwei356/seqkit/issues/19">#19</a>. <a href="http://bioinf.shenwei.me/seqkit/usage/#range">usage &amp; example</a>.</li>
<li>add new command <code>concate</code> for concatenating sequences with same ID from multiple files. <a href="http://bioinf.shenwei.me/seqkit/usage/#concate">usage &amp; example</a>.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.6.0">SeqKit v0.6.0</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.6.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.6.0/total.svg" /></a><ul>
<li>add new command <code>genautocomplete</code> to generate <strong>shell autocompletion</strong> script! (<a href="https://github.com/shenwei356/seqkit/issues/17">#17</a>)</li>
<li>add new command <code>seqkit dup</code> for duplicating sequences (<a href="https://github.com/shenwei356/seqkit/issues/16">#16</a>)</li>
<li><code>seqkit stats -a</code> does not show L50 which may brings confusion (<a href="https://github.com/shenwei356/seqkit/issues/15">#15</a>)</li>
<li><code>seqkit subseq --bed</code>: more robust for bad BED files</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.5">SeqKit v0.5.5</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.5"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.5.5/total.svg" /></a><ul>
<li>Increasing speed of reading <code>.gz</code> file by utilizing <code>gzip</code> (1.3X),
    it would be much faster if you installed <code>pigz</code> (2X).</li>
<li><strong><em>Fixing colorful output in Windows</em></strong></li>
<li><code>seqkit locate</code>: <strong><em>add flag <code>--gtf</code> and <code>--bed</code> to output GTF/BED6 format,
    so the result can be used in <code>seqkit subseq</code></em></strong>.</li>
<li><code>seqkit subseq</code>: fix bug of <code>--bed</code>, add checking coordinate.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.4">SeqKit v0.5.4</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.4"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.5.4/total.svg" /></a><ul>
<li><code>seqkit subseq --gtf</code>, add flag <code>--gtf-tag</code> to set tag that's outputted as sequence comment</li>
<li>fix <code>seqkit split</code> and <code>seqkit sample</code>: forget not to wrap sequence and quality in output for FASTQ format</li>
<li>compile with go1.8.1</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.3">SeqKit v0.5.3</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.3"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.5.3/total.svg" /></a><ul>
<li><code>seqkit grep</code>: fix bug when using <code>seqkit grep -r -f patternfile</code>:
  all records will be retrived due to failing to discarding the blank pattern (<code>""</code>). #11</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.2">SeqKit v0.5.2</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.5.2/total.svg" /></a><ul>
<li><code>seqkit stats -a</code> and <code>seqkit seq -g -G</code>: change default gap letters from '- ' to '- .'</li>
<li><code>seqkit subseq</code>: fix bug of range overflow when using <code>-d/--down-stream</code>
    or <code>-u/--up-stream</code> for retieving subseq using BED (<code>--beb</code>) or GTF (<code>--gtf</code>) file.</li>
<li><code>seqkit locate</code>: add flag <code>-G/--non-greedy</code>, non-greedy mode,
 faster but may miss motifs overlaping with others.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.1">SeqKit v0.5.1</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.5.1/total.svg" /></a><ul>
<li><code>seqkit restart</code>: fix bug of flag parsing</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.0">SeqKit v0.5.0</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.5.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.5.0/total.svg" /></a><ul>
<li><strong>new command <code>seqkit restart</code>, for resetting start position for circular genome</strong>.</li>
<li><code>seqkit sliding</code>: add flag <code>-g/--greedy</code>, exporting last subsequences even shorter than windows size.</li>
<li><code>seqkit seq</code>:<ul>
<li><strong>add flag <code>-m/--min-len</code> and <code>-M/--max-len</code> to filter sequences by length</strong>.</li>
<li>rename flag <code>-G/--gap-letter</code> to <code>-G/--gap-letters</code>.</li>
</ul>
</li>
<li><code>seqkit stat</code>:<ul>
<li>renamed to <code>seqkit stats</code>, don't worry, old name is still available as an alias.</li>
<li><strong>add new flag <code>-a/all</code>, for all statistics, including <code>sum_gap</code>, <code>N50</code>, and <code>L50</code></strong>.</li>
</ul>
</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.5">SeqKit v0.4.5</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.5"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.4.5/total.svg" /></a><ul>
<li><code>seqkit seq</code>: fix bug of failing to reverse quality of FASTQ sequence</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.4">SeqKit v0.4.4</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.4"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.4.4/total.svg" /></a><ul>
<li><code>seqkit locate</code>: fix bug of missing regular-expression motifs containing
 non-DNA characters (e.g., <code>ACT.{6,7}CGG</code>) from motif file (<code>-f</code>).</li>
<li>compiled with go v1.8.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.3">SeqKit v0.4.3</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.3"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.4.3/total.svg" /></a><ul>
<li>fix bug of <code>seqkit stat</code>: <code>min_len</code> always be <code>0</code> in versions: v0.4.0, v0.4.1, v0.4.2</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.2">SeqKit v0.4.2</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.4.2/total.svg" /></a><ul>
<li>fix header information of <code>seqkit subseq</code> when restriving up- and down-steam
sequences using GTF/BED file.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.1">SeqKit v0.4.1</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.4.1/total.svg" /></a><ul>
<li>enchancement: remove redudant regions for <code>seqkit locate</code>.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.0">SeqKit v0.4.0</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.4.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.4.0/total.svg" /></a><ul>
<li>fix bug of <code>seqkit locate</code>, e.g, only find two locations
(<code>1-4</code>, <code>7-10</code>, missing <code>4-7</code>) of <code>ACGA</code> in <code>ACGACGACGA</code>.</li>
<li>better output of <code>seqkit stat</code> for empty file.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.9">SeqKit v0.3.9</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.9"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.9/total.svg" /></a><ul>
<li>fix bug of region selection for blank sequences. affected commands include
<code>seqkit subseq --region</code>, <code>seqkit grep --region</code>, <code>seqkit split --by-region</code>.</li>
<li>compile with go1.8beta1.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.8.1">SeqKit v0.3.8.1</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.8.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.8.1/total.svg" /></a><ul>
<li>enhancement and bugfix of <code>seqkit common</code>: two or more same files allowed,
fix log information of number of extracted sequences in the first file.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.8">SeqKit v0.3.8</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.8"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.8/total.svg" /></a><ul>
<li>enhancement of <code>seqkit common</code>: better handling of files containing replicated sequences</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.7">SeqKit v0.3.7</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.7"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.7/total.svg" /></a><ul>
<li>fix bug in <code>seqkit split --by-id</code> when sequence ID contains invalid characters for system path.</li>
<li>add more flags validation for <code>seqkit replace</code>.</li>
<li>enhancement: raise error when key pattern matches multiple targes in cases of replacing with key-value files and more controls are added.</li>
<li>changes: do not wrap sequence and quality in output for FASTQ  format.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.6">SeqKit v0.3.6</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.6"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.6/total.svg" /></a><ul>
<li>add new feature for <code>seqkit grep</code>: new flag <code>-R</code> (<code>--region</code>) for specifying sequence region for searching.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.5">SeqKit v0.3.5</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.5"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.5/total.svg" /></a><ul>
<li>fig bug of <code>seqkit grep</code>: flag <code>-i</code> (<code>--ignore-case</code>) did not work when not using regular expression</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.4.1">SeqKit v0.3.4.1</a>
<a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.4.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.4.1/total.svg" /></a><ul>
<li>improve performance of reading (~10%) and writing (100%) gzip-compressed file
by using <code>github.com/klauspost/pgzip</code> package</li>
<li>add citation</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.4">SeqKit v0.3.4</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.4"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.4/total.svg" /></a><ul>
<li>bugfix: <code>seq</code> wrongly handles only the first one sequence file when multiple files given</li>
<li>new feature: <code>fx2tab</code> can output alphabet letters of a sequence by flag <code>-a</code> (<code>--alphabet</code>)</li>
<li>new feature: new flag <code>-K</code> (<code>--keep-key</code>) for <code>replace</code>,  when replacing
with key-value file, one can choose keeping the key as value or not.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.3">SeqKit v0.3.3</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.3"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.3/total.svg" /></a><ul>
<li>fix bug of <code>seqkit replace</code>, wrongly starting from 2 when using <code>{nr}</code>
  in <code>-r</code> (<code>--replacement</code>)</li>
<li>new feature: <code>seqkit replace</code> supports replacement symbols <code>{nr}</code> (record number)
  and <code>{kv}</code> (corresponding value of the key ($1) by key-value file)</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.2">SeqKit v0.3.2</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.2/total.svg" /></a><ul>
<li>fix bug of <code>seqkit split</code>, error when target file is in a directory.</li>
<li>improve performance of <code>seqkit spliding</code> for big sequences, and output
  last part even if it's shorter than window sze,
  output of FASTQ is also supported.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.1.1">SeqKit v0.3.1.1</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.1.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.1.1/total.svg" /></a><ul>
<li>compile with go1.7rc5, with <strong><em>higher performance and smaller size of binary file</em></strong></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.1">SeqKit v0.3.1</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.1/total.svg" /></a><ul>
<li>improve speed of <code>seqkit locate</code></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.0">SeqKit v0.3.0</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.3.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.3.0/total.svg" /></a><ul>
<li>use fork of github.com/brentp/xopen, using <code>zcat</code> for speedup of .gz file
  reading on *nix systems.</li>
<li>improve speed of parsing sequence ID when creating FASTA index</li>
<li>reduce memory usage of <code>seqkit subseq --gtf</code></li>
<li>fix bug of <code>seqkit subseq</code> when using flag <code>--id-ncbi</code></li>
<li>fix bug of <code>seqkit split</code>, outdir error</li>
<li>fix bug of <code>seqkit seq -p</code>, last base is wrongly failed to convert when
  sequence length is odd.</li>
<li>add "sum_len" result for output of <code>seqkit stat</code></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.9">seqkit v0.2.9</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.9"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.9/total.svg" /></a><ul>
<li>fix minor bug of <code>seqkit split</code> and <code>seqkit shuffle</code>,
  header name error due to improper use of pointer</li>
<li>add option <code>-O (--out-dir)</code> to <code>seqkit split</code></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.8">seqkit v0.2.8</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.8"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.8/total.svg" /></a><ul>
<li>improve speed of parsing sequence ID, not using regular expression for default <code>--id-regexp</code></li>
<li>improve speed of record outputing for small-size sequences</li>
<li>fix minor bug: <code>seqkit seq</code> for blank record</li>
<li>update benchmark result</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.7">seqkit v0.2.7</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.7"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.7/total.svg" /></a><ul>
<li><strong><em>reduce memory usage</em></strong> by optimize the outputing of sequences.
  detail: using <a href="https://godoc.org/github.com/shenwei356/util/byteutil#BufferedByteSliceWrapper"><code>BufferedByteSliceWrapper</code></a> to resuse bytes.Buffer.</li>
<li><strong><em>reduce memory usage and improve speed</em></strong> by using custom buffered
 reading mechanism, instead of using standard library <code>bufio</code>,
  which is slow for large genome sequence.</li>
<li>discard strategy of "buffer" and "chunk" of FASTA/Q records,
  just parse records one by one.</li>
<li>delete global flags <code>-c (--chunk-size)</code> and <code>-b (--buffer-size)</code>.</li>
<li>add function testing scripts</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.6">seqkit v0.2.6</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.6"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.6/total.svg" /></a><ul>
<li>fix bug of <code>seqkit subseq</code>: Inplace subseq method leaded to wrong result</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.5.1">seqkit v0.2.5.1</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.5.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.5.1/total.svg" /></a><ul>
<li>fix a bug of <code>seqkit subseq</code>: chromesome name was not be converting to lower case when using <code>--gtf</code> or <code>--bed</code></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.5">seqkit v0.2.5</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.5"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.5/total.svg" /></a><ul>
<li>fix a serious bug brought in <code>v0.2.3</code>, using unsafe method to convert <code>string</code> to <code>[]byte</code></li>
<li>add awk-like built-in variable of record number (<code>{NR}</code>) for <code>seqkit replace</code></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.4.1">seqkit v0.2.4.1</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.4.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.4.1/total.svg" /></a><ul>
<li>fix several bugs from library <code>bio</code>, affected situations:<ul>
<li>Locating patterns in sequences by pattern FASTA file: <code>seqkit locate -f</code></li>
<li>Reading FASTQ file with record of which the quality starts with <code>+</code></li>
</ul>
</li>
<li>add command <code>version</code></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.4">seqkit v0.2.4</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.4"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.4/total.svg" /></a><ul>
<li>add subcommand <code>head</code></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.3">seqkit v0.2.3</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.3"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.3/total.svg" /></a><ul>
<li>reduce memory occupation by avoid copy data when convert <code>string</code> to <code>[]byte</code></li>
<li>speedup reverse-complement by avoid repeatly calling functions</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.2">seqkit v0.2.2</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.2/total.svg" /></a><ul>
<li>reduce memory occupation of subcommands that use FASTA index</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.1">seqkit v0.2.1</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.1/total.svg" /></a><ul>
<li>improve performance of outputing.</li>
<li>fix bug of <code>seqkit seq -g</code> for FASTA fromat</li>
<li>some other minor fix of code and docs</li>
<li>update benchmark results</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.0">seqkit v0.2.0</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.2.0"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.2.0/total.svg" /></a><ul>
<li><strong><em>reduce memory usage of writing output</em></strong></li>
<li>fix bug of <code>subseq</code>, <code>shuffle</code>, <code>sort</code> when reading from stdin</li>
<li>reduce memory usage of <code>faidx</code></li>
<li>make validating sequences an optional option in <code>seq</code> command, it saves some time.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.9">seqkit v0.1.9</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.9"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.9/total.svg" /></a><ul>
<li>using custom FASTA index file extension: <code>.seqkit.fai</code></li>
<li>reducing memory usage of <code>sample --number --two-pass</code></li>
<li><strong><em>change default CPU number to 2 for multi-cpus computer, and 1 for single-CPU computer</em></strong></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.8">seqkit v0.1.8</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.8"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.8/total.svg" /></a><ul>
<li>add subcommand <code>rename</code> to rename duplicated IDs</li>
<li>add subcommand <code>faidx</code> to create FASTA index file</li>
<li><strong><em>utilize faidx to improve performance of <code>subseq</code></em></strong></li>
<li><em><code>shuffle</code>, <code>sort</code> and split support two-pass mode (by flag <code>-2</code>) with faidx to reduce memory usage.</em></li>
<li>document update</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.7">seqkit v0.1.7</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.7"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.7/total.svg" /></a><ul>
<li><strong><em>add support for (multi-line) FASTQ format</em></strong></li>
<li>update document, add technical details</li>
<li>rename subcommands <code>fa2tab</code> and <code>tab2fa</code> to <code>fx2tab</code> and <code>tab2fx</code></li>
<li>add subcommand <code>fq2fa</code></li>
<li>add column "seq_format" to <code>stat</code></li>
<li>add global flag <code>-b</code> (<code>--bufer-size</code>)</li>
<li>little change of flag in <code>subseq</code> and some other commands</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.6">seqkit v0.1.6</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.6"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.6/total.svg" /></a><ul>
<li>add subcommand <code>replace</code></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.5.2">seqkit v0.1.5.2</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.5.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.5.2/total.svg" /></a><ul>
<li>fix bug of <code>grep</code>, when not using flag <code>-r</code>, flag <code>-i</code> will not take effect.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.5.1">seqkit v0.1.5.1</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.5.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.5.1/total.svg" /></a><ul>
<li>fix result of <code>seqkit sample -n</code></li>
<li>fix benchmark script</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.5">seqkit v0.1.5</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.5"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.5/total.svg" /></a><ul>
<li>add global flag <code>--id-ncbi</code></li>
<li>add flag <code>-d</code> (<code>--dup-seqs-file</code>) and <code>-D</code> (<code>--dup-num-file</code>) for subcommand <code>rmdup</code></li>
<li>make using MD5 as an optional flag <code>-m</code> (<code>--md5</code>) in subcommand <code>rmdup</code> and <code>common</code></li>
<li>fix file name suffix of <code>seqkit split</code> result</li>
<li>minor modification of <code>sliding</code> output</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.4.1">seqkit v0.1.4.1</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.4.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.4.1/total.svg" /></a><ul>
<li>change alignment of <code>stat</code> output</li>
<li>preciser CPUs number control</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.4">seqkit v0.1.4</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.4"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.4/total.svg" /></a><ul>
<li>add subcommand <code>sort</code></li>
<li>improve subcommand <code>subseq</code>: supporting of getting subsequences by GTF and BED files</li>
<li>change name format of <code>sliding</code> result</li>
<li>prettier output of <code>stat</code></li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.3.1">seqkit v0.1.3.1</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.3.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.3.1/total.svg" /></a><ul>
<li>Performance improvement by reducing time of cleaning spaces</li>
<li>Document update</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.3">seqkit v0.1.3</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.3"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.3/total.svg" /></a><ul>
<li><strong>Further performance improvement</strong></li>
<li>Rename sub command <code>extract</code> to <code>grep</code></li>
<li>Change default value of flag <code>--threads</code> back CPU number of current device,
  change default value of flag <code>--chunk-size</code> back 10000 sequences.</li>
<li>Update benchmark</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.2">seqkit v0.1.2</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.2"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.2/total.svg" /></a><ul>
<li>Add flag <code>--dna2rna</code> and <code>--rna2dna</code> to subcommand <code>seq</code>.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1.1">seqkit v0.1.1</a>
  <a href="https://github.com/shenwei356/seqkit/releases/download/v0.1.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1.1/total.svg" /></a><ul>
<li><strong>5.5X speedup of FASTA file parsing</strong> by avoid using regular expression to remove spaces (<a href="https://github.com/shenwei356/bio/commit/2457b877cf1b8d79d05adb1a8952f2dff9046eaf">detail</a> ) and using slice indexing instead of map to validate letters (<a href="https://github.com/shenwei356/bio/commit/0f5912f6a3c6d737faacf9212f62d11c94e5044a">detail</a>)</li>
<li>Change default value of global flag <code>-- thread</code> to 1. Since most of the subcommands are I/O intensive,  For computation intensive jobs, like extract and locate, you may set a bigger value.</li>
<li>Change default value of global flag <code>--chunk-size</code> to 100.</li>
<li>Add subcommand <code>stat</code></li>
<li>Fix bug of failing to automatically detect alphabet when only one record in file.</li>
</ul>
</li>
<li><a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1">seqkit v0.1</a>
  <a href="https://github.com/shenwei356/seqkit/releases/tag/v0.1"><img alt="Github Releases (by Release)" src="https://img.shields.io/github/downloads/shenwei356/seqkit/v0.1/total.svg" /></a><ul>
<li>first release of seqkit</li>
</ul>
</li>
</ul>
<div id="disqus_thread"></div>

<script>

/**
*  RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
*  LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables*/
/*
var disqus_config = function () {
this.page.url = PAGE_URL;  // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');
s.src = '//seqkit.disqus.com/embed.js';
s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>

<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript></div>
        
        
    </div>

    <footer class="col-md-12 text-center">
        <hr>
        <p>
        <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p></small>
    </footer>

    <script src="../js/jquery-1.10.2.min.js"></script>
    <script src="../js/bootstrap-3.0.3.min.js"></script>
    <script src="../js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script>
    var base_url = '..';
    </script>
    <script data-main="../mkdocs/js/search.js" src="../mkdocs/js/require.js"></script>
    <script src="../js/base.js"></script>
    <script src="../search/main.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal">
                        <span aria-hidden="true">&times;</span>
                        <span class="sr-only">Close</span>
                    </button>
                    <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                </div>
                <div class="modal-body">
                    <p>
                        From here you can search these documents. Enter your search terms below.
                    </p>
                    <form role="form">
                        <div class="form-group">
                            <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                        </div>
                    </form>
                    <div id="mkdocs-search-results"></div>
                </div>
                <div class="modal-footer">
                </div>
            </div>
        </div>
    </div>

    </body>

</html>
